<!DOCTYPE html>
<html>
<!-- Created by GNU Texinfo 7.2, https://www.gnu.org/software/texinfo/ -->
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">

<!--
Copyright © 2026 Julien Y. Dutheil
SPDX-FileCopyrightText: 2026 Julien Y. Dutheil <jy.dutheil@gmail.com>

SPDX-License-Identifier: CC0-1.0
-->

<title>MafFilter Manual 1.4.0</title>

<meta name="description" content="MafFilter Manual 1.4.0">
<meta name="keywords" content="MafFilter Manual 1.4.0">
<meta name="resource-type" content="document">
<meta name="distribution" content="global">
<meta name="Generator" content="makeinfo">
<meta name="viewport" content="width=device-width,initial-scale=1">

<link href="#Top" rel="start" title="Top">
<link href="#SEC_Contents" rel="contents" title="Table of Contents">
<link href="dir.html#Top" rel="up" title="(dir)">
<link href="#Introduction" rel="next" title="Introduction">
<link href="dir.html#Top" rel="prev" title="(dir)">
<style type="text/css">
<!--
a.copiable-link {visibility: hidden; text-decoration: none; line-height: 0em}
div.example {margin-left: 3.2em}
span:hover a.copiable-link {visibility: visible}
table.cartouche {border-style: solid; border-radius: 0.5em}
ul.mark-bullet {list-style-type: disc}
ul.toc-numbered-mark {list-style: none}
-->
</style>
<link rel="stylesheet" type="text/css" href="http://www.w3.org/StyleSheets/Core/Steely">


</head>

<body lang="en">





<div class="top-level-extent" id="Top">
<div class="nav-panel">
<p>
Next: <a href="#Introduction" accesskey="n" rel="next">Introduction</a>, Previous: <a href="dir.html#Top" accesskey="p" rel="prev">(dir)</a>, Up: <a href="dir.html#Top" accesskey="u" rel="up">(dir)</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h1 class="top" id="The-MafFilter-Manual"><span>The MafFilter Manual<a class="copiable-link" href="#The-MafFilter-Manual"> &para;</a></span></h1>

<p>This is the manual of MafFilter, version 1.4.0.
</p>
<p>Copyright &copy; 2026 Julien Y. Dutheil
</p>



<div class="region-contents" id="SEC_Contents">
<h2 class="contents-heading">Table of Contents</h2>

<div class="contents">

<ul class="toc-numbered-mark">
  <li><a id="toc-Introduction" href="#Introduction">1 Introduction</a>
  <ul class="toc-numbered-mark">
    <li><a id="toc-Description-of-the-program" href="#Description">1.1 Description of the program</a></li>
    <li><a id="toc-How-to-run-the-program" href="#Run">1.2 How to run the program</a></li>
    <li><a id="toc-General-options" href="#General">1.3 General options</a></li>
  </ul></li>
  <li><a id="toc-Filters" href="#Filters">2 Filters</a>
  <ul class="toc-numbered-mark">
    <li><a id="toc-Extracting-data" href="#Extracting">2.1 Extracting data</a>
    <ul class="toc-numbered-mark">
      <li><a id="toc-Subset_003a-remove-sequences-from-blocks-by-species-name" href="#Subset">2.1.1 Subset: remove sequences from blocks by species name</a></li>
      <li><a id="toc-SelectOrphans_003a-keep-only-block-with-the-specified-species" href="#SelectOrphans">2.1.2 SelectOrphans: keep only block with the specified species</a></li>
      <li><a id="toc-Merging_003a-merge-synthenic-blocks" href="#Merge">2.1.3 Merging: merge synthenic blocks</a></li>
      <li><a id="toc-Concatenating_003a-merge-consecutive-blocks-up-to-a-certain-size" href="#Concatenate">2.1.4 Concatenating: merge consecutive blocks up to a certain size</a></li>
      <li><a id="toc-OrderFilter_003a-check-and-eventually-discard-unordered-or-overlapping-blocks" href="#OrderFilter">2.1.5 OrderFilter: check and eventually discard unordered or overlapping blocks</a></li>
      <li><a id="toc-Extract-features-from-the-alignment" href="#ExtractFeature">2.1.6 Extract features from the alignment</a></li>
      <li><a id="toc-Extract-blocks-for-a-given-chromosome_002fcontig_002fscaffold" href="#SelectChr">2.1.7 Extract blocks for a given chromosome/contig/scaffold</a></li>
      <li><a id="toc-Rename-chromosomes-in-each-block" href="#RenameChr">2.1.8 Rename chromosomes in each block</a></li>
      <li><a id="toc-Split-alignment-blocks-into-smaller-parts" href="#WindowSplit">2.1.9 Split alignment blocks into smaller parts</a></li>
    </ul></li>
    <li><a id="toc-Cleaning-alignment-blocks" href="#Cleaning">2.2 Cleaning alignment blocks</a>
    <ul class="toc-numbered-mark">
      <li><a id="toc-Minumum-block-length" href="#MinBlockLength">2.2.1 Minumum block length</a></li>
      <li><a id="toc-Minumum-block-size" href="#MinBlockSize">2.2.2 Minumum block size</a></li>
      <li><a id="toc-Minumum-alignment-score" href="#MinAlignmentScore">2.2.3 Minumum alignment score</a></li>
      <li><a id="toc-Exclude-full-gap-columns" href="#XFullGap">2.2.4 Exclude full gap columns</a></li>
      <li><a id="toc-Remove-empty-sequences" href="#RemoveEmptySequences">2.2.5 Remove empty sequences</a></li>
      <li><a id="toc-Alignment-filtering" href="#AlnFilter">2.2.6 Alignment filtering</a></li>
      <li><a id="toc-Alignment-filtering-2" href="#AlnFilter2">2.2.7 Alignment filtering 2</a></li>
      <li><a id="toc-Entropy-filtering" href="#EntropyFilter">2.2.8 Entropy filtering</a></li>
      <li><a id="toc-Masked-sequences-filtering" href="#MaskFilter">2.2.9 Masked sequences filtering</a></li>
      <li><a id="toc-Quality-filtering" href="#QualFilter">2.2.10 Quality filtering</a></li>
      <li><a id="toc-Remove-sequences-from-features-from-the-alignment" href="#FeatureFilter">2.2.11 Remove sequences from features from the alignment</a></li>
    </ul></li>
    <li><a id="toc-Statistical-analysis" href="#Analyzing">2.3 Statistical analysis</a>
    <ul class="toc-numbered-mark">
      <li><a id="toc-Descriptive-statistics" href="#Descriptive">2.3.1 Descriptive statistics</a>
      <ul class="toc-numbered-mark">
        <li><a id="toc-Size-of-alignment-blocks" href="#BlockSize">2.3.1.1 Size of alignment blocks</a></li>
        <li><a id="toc-Length-of-alignment-blocks" href="#BlockLength">2.3.1.2 Length of alignment blocks</a></li>
        <li><a id="toc-Length-of-sequence-for-a-species" href="#SequenceLength">2.3.1.3 Length of sequence for a species</a></li>
        <li><a id="toc-Alignment-score" href="#AlnScore">2.3.1.4 Alignment score</a></li>
        <li><a id="toc-Character-frequencies" href="#BlockCounts">2.3.1.5 Character frequencies</a></li>
        <li><a id="toc-Pairwise-divergence" href="#PairwiseDivergence">2.3.1.6 Pairwise divergence</a></li>
        <li><a id="toc-Site-statistics" href="#SiteStatistics">2.3.1.7 Site statistics</a></li>
        <li><a id="toc-Counts-of-parsimony-informative-sites-for-a-species-quartet" href="#FourSpeciesSitePatternCounts">2.3.1.8 Counts of parsimony informative sites for a species quartet</a></li>
        <li><a id="toc-Site-frequency-spectrum" href="#SiteFrequencySpectrum">2.3.1.9 Site frequency spectrum</a></li>
        <li><a id="toc-Polymorphism-statistics" href="#PolymorphismStatistics">2.3.1.10 Polymorphism statistics</a></li>
        <li><a id="toc-Fst-statistics" href="#FstStatistics">2.3.1.11 Fst statistics</a></li>
        <li><a id="toc-Sequence-diversity-statistics" href="#DiversityStatistics">2.3.1.12 Sequence diversity statistics</a></li>
        <li><a id="toc-Number-of-sequence-clusters_002e" href="#CountClusters">2.3.1.13 Number of sequence clusters.</a></li>
        <li><a id="toc-Fit-a-substitution-model-and-estimate-parameters" href="#ModelFit">2.3.1.14 Fit a substitution model and estimate parameters</a></li>
      </ul></li>
      <li><a id="toc-Phylogenetics" href="#Phylogenetics">2.3.2 Phylogenetics</a>
      <ul class="toc-numbered-mark">
        <li><a id="toc-Distance-matrix-estimation" href="#DistanceEstimation">2.3.2.1 Distance matrix estimation</a></li>
        <li><a id="toc-Hierarchical-clustering-based-on-distances" href="#DistanceBasedPhylogeny">2.3.2.2 Hierarchical clustering based on distances</a></li>
        <li><a id="toc-Call-an-external-phylogenetic-software-to-reconstruct-tree" href="#ExternalTreeBuilding">2.3.2.3 Call an external phylogenetic software to reconstruct tree</a></li>
        <li><a id="toc-Root-trees" href="#NewOutgroup">2.3.2.4 Root trees</a></li>
        <li><a id="toc-Remove-a-species-from-a-tree" href="#DropSpecies">2.3.2.5 Remove a species from a tree</a></li>
        <li><a id="toc-Filter-blocks-based-on-their-associated-trees" href="#TreeFilter">2.3.2.6 Filter blocks based on their associated trees</a></li>
      </ul></li>
    </ul></li>
    <li><a id="toc-Exporting-blocks-and-data" href="#Exporting">2.4 Exporting blocks and data</a>
    <ul class="toc-numbered-mark">
      <li><a id="toc-Write-alignment-blocks-to-a-MAF-file" href="#Output">2.4.1 Write alignment blocks to a MAF file</a></li>
      <li><a id="toc-Write-alignment-blocks-to-an-external-alignment-file_002e" href="#OutputAlignments">2.4.2 Write alignment blocks to an external alignment file.</a></li>
      <li><a id="toc-Export-given-positions-into-a-table-format_002e" href="#OutputAsTable">2.4.3 Export given positions into a table format.</a></li>
      <li><a id="toc-Call-SNPs-from-alignment-blocks-and-export-to-VCF-file" href="#VcfOutput">2.4.4 Call SNPs from alignment blocks and export to VCF file</a></li>
      <li><a id="toc-Export-segregating-sites-to-MSMC-format" href="#MsmcOutput">2.4.5 Export segregating sites to MSMC format</a></li>
      <li><a id="toc-Export-segregating-sites-to-PLINK-format" href="#PlinkOutput">2.4.6 Export segregating sites to PLINK format</a></li>
      <li><a id="toc-Output-coordinates-of-a-set-of-species-for-each-block_002e" href="#OutputCoordinates">2.4.7 Output coordinates of a set of species for each block.</a></li>
      <li><a id="toc-Write-trees-to-a-file_002e" href="#OutputTrees">2.4.8 Write trees to a file.</a></li>
      <li><a id="toc-Write-distance-matrices-to-a-file_002e" href="#OutputDistanceMatrices">2.4.9 Write distance matrices to a file.</a></li>
    </ul></li>
    <li><a id="toc-Miscelaneous-filters" href="#Miscelaneous">2.5 Miscelaneous filters</a>
    <ul class="toc-numbered-mark">
      <li><a id="toc-Convert-coordinates-from-one-genome-to-another" href="#LiftOver">2.5.1 Convert coordinates from one genome to another</a></li>
      <li><a id="toc-Call-an-external-program-on-each-block_002e" href="#SystemCall">2.5.2 Call an external program on each block.</a></li>
    </ul></li>
  </ul></li>
</ul>
</div>
</div>
<hr>
<div class="chapter-level-extent" id="Introduction">
<div class="nav-panel">
<p>
Next: <a href="#Filters" accesskey="n" rel="next">Filters</a>, Previous: <a href="#Top" accesskey="p" rel="prev">The MafFilter Manual</a>, Up: <a href="#Top" accesskey="u" rel="up">The MafFilter Manual</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h2 class="chapter" id="Introduction-1"><span>1 Introduction<a class="copiable-link" href="#Introduction-1"> &para;</a></span></h2>

<p>MafFilter is a program for post-processing genome alignments in the Multiple Alignment Files (MAF) format.
MAF files consist of a series of &quot;blocks&quot;, displaying the alignment of subsequences of the original genomes.
In a typical MAF file, such blocks represent synteny blocks.
See the UCSC website for a detailed description of the file format.
</p>
<p>How to obtain MAF files is not covered in this manual. MAF files are typically generated by packages such as BlastZ, MultiZ, Last or Galaxy, and you should refer to the manual of these resources for more detailed help.
These packages also contain tools for manipulating MAF files, which you might find useful in combination to the program described here.
</p>
<p>Note that several examples are provided along with the source code of the program, and can serve as good training starts.
This manual intends to provide an exhaustive description of the options used in these examples, as well as to provide general information about the way the program works and its underlying algorithms.
</p>

         

<ul class="mini-toc">
<li><a href="#Description" accesskey="1">Description of the program</a></li>
<li><a href="#Run" accesskey="2">How to run the program</a></li>
<li><a href="#General" accesskey="3">General options</a></li>
</ul>
<hr>
<div class="section-level-extent" id="Description">
<div class="nav-panel">
<p>
Next: <a href="#Run" accesskey="n" rel="next">How to run the program</a>, Previous: <a href="#Introduction" accesskey="p" rel="prev">Introduction</a>, Up: <a href="#Introduction" accesskey="u" rel="up">Introduction</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="Description-of-the-program"><span>1.1 Description of the program<a class="copiable-link" href="#Description-of-the-program"> &para;</a></span></h3>

<p>The main goal of MafFilter is to apply a series of &quot;filters&quot; to the original input file.
A &ldquo;filter&rdquo; can be seen as a small program that takes as input one or several maf blocks, and outputs one or several maf block, after performing some actions on them:
</p>
<pre class="verbatim">                         +--------+
   Alignment block(s)    |        |   Alignment block(s)
------------------------&gt;+ Filter +------------------------&gt;
                         |        |
                         +--------+
</pre>
<p>We can roughly distinguish four types of filters:
</p>
<ul class="itemize mark-bullet">
<li>Filters that take one block as input, and output one single block, possibly after removing some sequences, or even the full block if it did not match a certain criterion,
</li><li>Filters that take one block as input, and split them into several smaller blocks, for instance of a fixed size,
</li><li>Filters that take several blocks as input and output a larger one, typically after concatenating the input blocks,
</li><li>Filters that do not modify the input block and simply forward it, yet after having performed some computations, for instance calculating various statistics.
</li></ul>

<p>A run of MafFilter typically consists in a workflow of filters, applied in a serial way, the output of one filter being taken as input as the next one.
In the current version of MafFilter, only linear workflows are possible, without branching:
</p>
<pre class="verbatim">    +----------+    +----------+    +----------+
    |          |    |          |    |          |
---&gt;+ Filter 1 +---&gt;+ Filter 2 +---&gt;+ Filter 3 +---&gt; Etc.
    |          |    |          |    |          |
    +----------+    +----------+    +----------+
</pre>
<p>Branching processes can however be achieved by outputing the alignment blocks at the branching point, and run various instances of MafFilter on the resulting files:
</p>
<pre class="verbatim">     Run 1
    +----------+    +----------+    +----------+
    |          |    |          |    |          |
---&gt;+ Filter 1 +---&gt;+ Filter 2 +---&gt;+ Filter 3 +---&gt; Etc.
    |          |    |          |    |          |
    +----------+    +----+-----+    +----------+
                         |
                         |    +------------+
                         |    |            |
     Run 2               +---&gt;+ Filter 2.1 +---&gt; Etc.
                              |            |
                              +------------+
</pre>
<p>Some filters allow to generate additional data associated to each alignment block, such as phylogenetic trees.
Such data cannot be written to file together with the blocks, as the MAF format does not support this.
These data are however stored in memory by MafFilter together with the block, as &ldquo;extra-data&rdquo;.
Extra-data are associated with a special tag which is used for refering to them, including for writting them in appropriate output files.
</p>
<p>The next section describes how to specify the series of filters to be applied, and the next chapter provides an exhaustive list of available filters.
</p>

<hr>
</div>
<div class="section-level-extent" id="Run">
<div class="nav-panel">
<p>
Next: <a href="#General" accesskey="n" rel="next">General options</a>, Previous: <a href="#Description" accesskey="p" rel="prev">Description of the program</a>, Up: <a href="#Introduction" accesskey="u" rel="up">Introduction</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="How-to-run-the-program"><span>1.2 How to run the program<a class="copiable-link" href="#How-to-run-the-program"> &para;</a></span></h3>

<p>MafFilter is a program written in C++ with the Bio++ libraries. It is a command line program (that is, without graphical user interface) which uses the Bio++ Options (BppO) syntax. 
This syntax is of type &lsquo;<samp class="samp">keys</samp>&rsquo;=&lsquo;<samp class="samp">values</samp>&rsquo;, which can be passed directly as arguments of the maffilter program
</p><pre class="verbatim">maffilter option1=value1 option2=value2 ... optionN=valueN
</pre><p>or more conveniently, stored in an &ldquo;options&rdquo; file (specified with param=option_file), or a combination of both:
</p><pre class="verbatim">maffilter param=option_file
maffilter param=option_file option1=value1 option2=value2
</pre><p>See the Bio++ Program Suite (BppSuite) manual for more information.
</p>
<p>MafFilter typically takes as input a MAF file and a series of commands which determine the output file(s) that will be written and where.
Additional input files might be specified, such as gene annotation files (typically in GFF/GTF format).
Here is the structure of a minimal option file:
</p>
<pre class="verbatim">DATA=example.file //A user-defined variable, representing the input maf file, without extension
input.file=$(DATA).maf.gz  //Input maf file, gzipped
input.file.compression=gzip
output.log=$(DATA).maffilter.log //Output log file
maf.filter= //A comma-separated list of filters, here empty (the programs only read the input file and exits).
</pre><p>The list of filters can be split on multiple lines for convenience, providing each line is escaped by a &rsquo;\&rsquo; at the end to indicate that the next line is a continuation of the current one:
</p><pre class="verbatim">maf.filter=Filter1,\
           Filter2,\
           Filter3
</pre>
<p>Please refer to the Bio++ Program Suite manual for more details, including the use of variables, priority of option values, etc.
</p>

<hr>
</div>
<div class="section-level-extent" id="General">
<div class="nav-panel">
<p>
Previous: <a href="#Run" accesskey="p" rel="prev">How to run the program</a>, Up: <a href="#Introduction" accesskey="u" rel="up">Introduction</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="General-options"><span>1.3 General options<a class="copiable-link" href="#General-options"> &para;</a></span></h3>

<h3 class="heading" id="Synopsis_003a"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">input.file=mydata.maf.gz
input.file.compression=gzip
input.format=Maf
output.log=mydata.maffilter.log
...
</pre></div>

<p>or
</p><div class="example">
<pre class="example-preformatted">input.file=mydata.fasta
input.file.compression=none
input.format=Fasta(zero_based=yes)
output.log=mydata.maffilter.log
...
</pre></div>
</td></tr></table>

<h3 class="heading" id="Detailed-description-of-commands_003a"><span>Detailed description of commands:<a class="copiable-link" href="#Detailed-description-of-commands_003a"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">input.file={path}</code></dt>
<dd><p>The MAF file to parse
</p>
</dd>
<dt><code class="command">input.file.compression=[none|zip|gzip|bzip2]</code></dt>
<dd><p>The compression format of the input file.
</p>
</dd>
<dt><code class="command">input.format={Maf|Fasta}</code></dt>
<dd><p>MafFilter takes by default a MAF file. It is, however, possible to use a fasta file as input,
which will lead MafFilter to run in &quot;single sequence&quot; mode.
In this mode, each sequence of the original file is considered as a distinct block.
This mode might be useful to extract annotations and estimate simple statistics.
Fasta files can be compressed in the same way as MAF files.
The sequence names are expected to be of the form Species:Chromosome:start:strand:length, with possible empty fields.
Coordinates are expected to be 0-based (that is, the first nucleotide is position 0). However, passing the <code class="command">zero_based=no</code> argument to the format name will assume that the first nucleotide is position 1.
</p>
</dd>
<dt><code class="command">input.dots={error|as_gaps|as_unresolved}</code></dt>
<dd><p>If set to <code class="command">as_gaps</code>, dotted characters in sequences will be converted to gaps and treated as such in any further step of the analysis, including output to files. If set to <code class="command">as_unresolved</code>, dots will be converted to &rsquo;N&rsquo;.
</p>
</dd>
<dt><code class="command">input.check_sequence_size={boolean}</code></dt>
<dd><p>Tell if sequence sizes are consistent (default: true). An error will be returned in case of inconsistency.
Desabling size check can be useful in combination with input.dots=as_gaps. In case of inconsistency, a warning will be displayed unless verbose is set to false.
</p>
</dd>
<dt><code class="command">output.log={path}</code></dt>
<dd><p>The file where to write log messages.
</p>
</dd>
<dt><code class="command">maf.filter={Filter1, Filter2, ...}</code></dt>
<dd><p>The main command, specifying the list of filter commands to apply, in the given order (that is, the input of filter 2 will be the output of filter 1).
Filters take a series of arguments, specified within parentheses: <code class="command">Filter(arg1=opt1, arg2=opt2, etc)</code>.
The next section present all available filters and their corresponding arguments.
</p></dd>
</dl>


<hr>
</div>
</div>
<div class="chapter-level-extent" id="Filters">
<div class="nav-panel">
<p>
Previous: <a href="#Introduction" accesskey="p" rel="prev">Introduction</a>, Up: <a href="#Top" accesskey="u" rel="up">The MafFilter Manual</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h2 class="chapter" id="Filters-1"><span>2 Filters<a class="copiable-link" href="#Filters-1"> &para;</a></span></h2>



<ul class="mini-toc">
<li><a href="#Extracting" accesskey="1">Extracting data</a></li>
<li><a href="#Cleaning" accesskey="2">Cleaning alignment blocks</a></li>
<li><a href="#Analyzing" accesskey="3">Statistical analysis</a></li>
<li><a href="#Exporting" accesskey="4">Exporting blocks and data</a></li>
<li><a href="#Miscelaneous" accesskey="5">Miscelaneous filters</a></li>
</ul>
<hr>
<div class="section-level-extent" id="Extracting">
<div class="nav-panel">
<p>
Next: <a href="#Cleaning" accesskey="n" rel="next">Cleaning alignment blocks</a>, Previous: <a href="#Filters" accesskey="p" rel="prev">Filters</a>, Up: <a href="#Filters" accesskey="u" rel="up">Filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="Extracting-data"><span>2.1 Extracting data<a class="copiable-link" href="#Extracting-data"> &para;</a></span></h3>



<ul class="mini-toc">
<li><a href="#Subset" accesskey="1">Subset: remove sequences from blocks by species name</a></li>
<li><a href="#SelectOrphans" accesskey="2">SelectOrphans: keep only block with the specified species</a></li>
<li><a href="#Merge" accesskey="3">Merging: merge synthenic blocks</a></li>
<li><a href="#Concatenate" accesskey="4">Concatenating: merge consecutive blocks up to a certain size</a></li>
<li><a href="#OrderFilter" accesskey="5">OrderFilter: check and eventually discard unordered or overlapping blocks</a></li>
<li><a href="#ExtractFeature" accesskey="6">Extract features from the alignment</a></li>
<li><a href="#SelectChr" accesskey="7">Extract blocks for a given chromosome/contig/scaffold</a></li>
<li><a href="#RenameChr" accesskey="8">Rename chromosomes in each block</a></li>
<li><a href="#WindowSplit" accesskey="9">Split alignment blocks into smaller parts</a></li>
</ul>
<hr>
<div class="subsection-level-extent" id="Subset">
<div class="nav-panel">
<p>
Next: <a href="#SelectOrphans" accesskey="n" rel="next">SelectOrphans: keep only block with the specified species</a>, Previous: <a href="#Extracting" accesskey="p" rel="prev">Extracting data</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Subset_003a-remove-sequences-from-blocks-by-species-name"><span>2.1.1 Subset: remove sequences from blocks by species name<a class="copiable-link" href="#Subset_003a-remove-sequences-from-blocks-by-species-name"> &para;</a></span></h4>

<p>The <code class="command">Subset</code> filter allows to filter sequences for each block by keeping only a selection of species.
</p>
<h3 class="heading" id="Synopsis_003a-1"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-1"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    Subset(                                 \
        species=(species1, species2, etc),  \
        strict=yes,                         \
        keep=no,                            \
        remove_duplicates=yes),             \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a"><span>Arguments:<a class="copiable-link" href="#Arguments_003a"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. Only sequences corresponding to the selected species will be kept.
Block without sequences from the selection will be discarded.
</p>
</dd>
<dt><code class="command">strict=yes/no</code></dt>
<dd><p>If set to yes, only blocks containing at least one sequence from each species in the selection will be kept.
</p>        
</dd>
<dt><code class="command">keep=yes/no</code></dt>
<dd><p>If strict is set to yes, tells if additional sequences should be kept in the block, or remove so that the output blocks only contain species from the selection.
</p>        
</dd>
<dt><code class="command">remove_duplicates=yes/no</code></dt>
<dd><p>If set to yes, only keep blocks for which only one sequence for each species is present. Blocks with paralogous sequences in at least one species will therefore be discarded. 
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="SelectOrphans">
<div class="nav-panel">
<p>
Next: <a href="#Merge" accesskey="n" rel="next">Merging: merge synthenic blocks</a>, Previous: <a href="#Subset" accesskey="p" rel="prev">Subset: remove sequences from blocks by species name</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="SelectOrphans_003a-keep-only-block-with-the-specified-species"><span>2.1.2 SelectOrphans: keep only block with the specified species<a class="copiable-link" href="#SelectOrphans_003a-keep-only-block-with-the-specified-species"> &para;</a></span></h4>

<p>The <code class="command">SelectOrphans</code> filter allows to discard all block with species out of the specified range.
This is typically used to extract blocks with no homolog (&quot;orphan&quot;).
</p>
<h3 class="heading" id="Synopsis_003a-2"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-2"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    SelectOrphans(                          \
        species=(species1, species2, etc),  \
        strict=yes,                         \
        remove_duplicates=yes),             \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-1"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-1"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species.
All block containing at least one species outside this list will be discarded.
</p>
</dd>
<dt><code class="command">strict=yes/no</code></dt>
<dd><p>If set to yes, only block containing at least one sequence from each species in the selection will be kept.
</p>        
</dd>
<dt><code class="command">keep=yes/no</code></dt>
<dd><p>If strict is set to yes, tells if additional sequences should be kept in the block, or remove so that the output blocks only contain species from the selection.
</p>        
</dd>
<dt><code class="command">remove_duplicates=yes/no</code></dt>
<dd><p>If set to yes, only keep blocks for which only one sequence for each species is present.
Blocks with paralogous sequences in at least one species will therefore be discarded. 
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="Merge">
<div class="nav-panel">
<p>
Next: <a href="#Concatenate" accesskey="n" rel="next">Concatenating: merge consecutive blocks up to a certain size</a>, Previous: <a href="#SelectOrphans" accesskey="p" rel="prev">SelectOrphans: keep only block with the specified species</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Merging_003a-merge-synthenic-blocks"><span>2.1.3 Merging: merge synthenic blocks<a class="copiable-link" href="#Merging_003a-merge-synthenic-blocks"> &para;</a></span></h4>

<p>The <code class="command">Merge</code> filter fuses consecutive blocks based on their coordinates.
Blocks will be fused only if the sequences from a specified set of species can be considered as syntenic.
Two sequences are considered as syntenic if:
</p><ul class="itemize mark-bullet">
<li>they are from the same species
</li><li>they are from the same chromosome
</li><li>they are from the same strand
</li><li>they are distant by less than a defined threshold (the start of sequence 2 - the end of sequence 1 &lt; threshold in bp).
</li></ul>

<h3 class="heading" id="Synopsis_003a-3"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-3"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    Merge(                                  \
        species=(species1, species2, etc),  \
        dist_max=0,                         \
        ignore_chr=(Random,Unknown),        \
        rename_chimeric_chromosomes=yes),   \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-2"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-2"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species.
These species only will be used for the synteny check.
Other species will be concatenated in any situation, and their coordinates removed.
</p>
</dd>
<dt><code class="command">dist_max={int}</code></dt>
<dd><p>Maximum distance (in nt) to consider sequences as syntenic.
If dist_max is greater than 0, sequences will be filled with &rsquo;N&rsquo; to preserve coordinates when they are fused.
</p>
</dd>
<dt><code class="command">ignore_chr={none | (list)}</code></dt>
<dd><p>An optional parenthetic list of chromosomes to ignore (typically &quot;Unknown&quot;, or &quot;Random&quot;, etc.).
Sequences annotated with such chromosomes will not be checked for synteny and the corresponding block will not be fused.
</p>
</dd>
<dt><code class="command">rename_chimeric_chromosomes={boolean}</code></dt>
<dd><p>When distinct chromosomes are merged (in non specified species), their names are concatenated. If this option is set to &quot;yes&quot;,
then merged chromosomes will be renamed as &quot;chimtigXXX&quot; where XXX is a number auto-incremented for each species.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="Concatenate">
<div class="nav-panel">
<p>
Next: <a href="#OrderFilter" accesskey="n" rel="next">OrderFilter: check and eventually discard unordered or overlapping blocks</a>, Previous: <a href="#Merge" accesskey="p" rel="prev">Merging: merge synthenic blocks</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Concatenating_003a-merge-consecutive-blocks-up-to-a-certain-size"><span>2.1.4 Concatenating: merge consecutive blocks up to a certain size<a class="copiable-link" href="#Concatenating_003a-merge-consecutive-blocks-up-to-a-certain-size"> &para;</a></span></h4>

<p>The <code class="command">Concatenate</code> filter fuses consecutive blocks until the concatenated block reaches a minimal size.
</p>
<h3 class="heading" id="Synopsis_003a-4"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-4"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    Concatenate(minimum_size=10000, ref_species=Hsapiens), \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-3"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-3"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">minimum_size={int}</code></dt>
<dd><p>The minimum size for the blocks to be reached.
</p>
</dd>
<dt><code class="command">ref_species={string}</code></dt>
<dd><p>If given, only blocks with identical chromosome tags in the reference species will be merged.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="OrderFilter">
<div class="nav-panel">
<p>
Next: <a href="#ExtractFeature" accesskey="n" rel="next">Extract features from the alignment</a>, Previous: <a href="#Concatenate" accesskey="p" rel="prev">Concatenating: merge consecutive blocks up to a certain size</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="OrderFilter_003a-check-and-eventually-discard-unordered-or-overlapping-blocks"><span>2.1.5 OrderFilter: check and eventually discard unordered or overlapping blocks<a class="copiable-link" href="#OrderFilter_003a-check-and-eventually-discard-unordered-or-overlapping-blocks"> &para;</a></span></h4>

<p>The <code class="command">OrderFilter</code> filter check if blocks are sorted according to previous block.
</p>
<h3 class="heading" id="Synopsis_003a-5"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-5"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    OrderFilter(                            \
        reference=ref_genome,               \
        do_unsorted=discard,                \
        do_overlapping=discard),            \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-4"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-4"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">reference={string}</code></dt>
<dd><p>The sequence for which coordinates should be checked. Block where the reference sequence is absent will be considered as ordered.
</p>
<p>Important note: the alignment is supposed to be projected according to the reference sequence. Using the filter on another species might lead to unexpected behavior, in particular as the the reference species is supposed to be mapped on the + strand only.
</p>
</dd>
<dt><code class="command">do_unsorted={none|discard|error}</code></dt>
<dd><p>Action to perform in case an unsorted block is found (that is, when the right coordinate of the current block is before the left coordinate of the previous one).
If &rsquo;discard&rsquo;, the block will be removed. If &rsquo;error&rsquo;, the maffilter with stop with an error.
</p>
</dd>
<dt><code class="command">do_overlapping={none|discard|error}</code></dt>
<dd><p>Action to perform in case an ovelrapping block is found (that is, when the left coordinate of the current block is before the right coordinate of the previous one, and the right coordinate of the current block is not before the left coordinate of the previous one).
If &rsquo;discard&rsquo;, the block will be removed. If &rsquo;error&rsquo;, the maffilter with stop with an error.
</p></dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="ExtractFeature">
<div class="nav-panel">
<p>
Next: <a href="#SelectChr" accesskey="n" rel="next">Extract blocks for a given chromosome/contig/scaffold</a>, Previous: <a href="#OrderFilter" accesskey="p" rel="prev">OrderFilter: check and eventually discard unordered or overlapping blocks</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Extract-features-from-the-alignment"><span>2.1.6 Extract features from the alignment<a class="copiable-link" href="#Extract-features-from-the-alignment"> &para;</a></span></h4>

<p>The <code class="command">ExtractFeature</code> extracts parts of the alignment corresponding to certain features.
The features to extract are typically provided via a GFF file. 
</p>
<h3 class="heading" id="Synopsis_003a-6"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-6"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...]
    ExtractFeature(                         \
        ref_species=species1,               \
        feature.file=species1.gff3.gz,      \
        feature.file.compression=gzip,      \
        feature.format=GFF,                 \
        feature.type=mRNA,                  \
        compression=gzip,                   \
        complete=yes,                       \
        ignore_strand=no),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-5"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-5"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">ref_species={string}</code></dt>
<dd><p>The name of the species for which the coordinates of the features are provided.
</p>
</dd>
<dt><code class="command">feature.file={path}</code></dt>
<dd><p>The file where the features are described.
</p>
</dd>
<dt><code class="command">feature.file.compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for the feature file.
</p>
</dd>
<dt><code class="command">feature.format={GFF|GTF|BedGraph}</code></dt>
<dd><p>Format for the feature file, currently GFF (v3.0), GTF or BedGraph.
</p>
</dd>
<dt><code class="command">feature.type={all|(list)}</code></dt>
<dd><p>Specifies which type of feature should be extracted from the feature file, if several are available.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
<dt><code class="command">complete={yes|no}</code></dt>
<dd><p>Tell if only complete features should be extracted or if partial extraction is allowed.
</p>
</dd>
<dt><code class="command">ignore_strand={yes|no}</code></dt>
<dd><p>If yes, features will be extracted &ldquo;as is&rdquo;. Otherwise, the invert-complement sequence will be returned if it is located on the negative strand.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="SelectChr">
<div class="nav-panel">
<p>
Next: <a href="#RenameChr" accesskey="n" rel="next">Rename chromosomes in each block</a>, Previous: <a href="#ExtractFeature" accesskey="p" rel="prev">Extract features from the alignment</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Extract-blocks-for-a-given-chromosome_002fcontig_002fscaffold"><span>2.1.7 Extract blocks for a given chromosome/contig/scaffold<a class="copiable-link" href="#Extract-blocks-for-a-given-chromosome_002fcontig_002fscaffold"> &para;</a></span></h4>

<p>The <code class="command">SelectChr</code> filter keeps only block from a given set of chromosomes (or contig or scaffold...) for a given reference species.
</p>
<h3 class="heading" id="Synopsis_003a-7"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-7"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    SelectChr(                              \
        ref_species=species1,               \
        chromosome=(chr01,chr02)),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-6"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-6"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">ref_species={string}</code></dt>
<dd><p>The species name for which chromosomes should be extracted.
</p>
</dd>
<dt><code class="command">chromosome={(list)}</code></dt>
<dd><p>The name of the chromosome(s)/contig(s)/scaffold(s) to be filtered, separated by commas.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="RenameChr">
<div class="nav-panel">
<p>
Next: <a href="#WindowSplit" accesskey="n" rel="next">Split alignment blocks into smaller parts</a>, Previous: <a href="#SelectChr" accesskey="p" rel="prev">Extract blocks for a given chromosome/contig/scaffold</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Rename-chromosomes-in-each-block"><span>2.1.8 Rename chromosomes in each block<a class="copiable-link" href="#Rename-chromosomes-in-each-block"> &para;</a></span></h4>

<p>The <code class="command">RenameChr</code> update the chromosome names according to a translation table.
</p>
<h3 class="heading" id="Synopsis_003a-8"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-8"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    RenameChr(                              \
        translation_file=chr_tln.csv,       \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-7"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-7"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">translation_file={path}</code></dt>
<dd><p>File path toward a translation file. The translation file should contain two columns, separated by commas, and no header.
The first column contains the current chromosome names, and the second column contains the new name.
Chromosome not mentionned in the table will be left untranslated. Additional columns in the table will be ignored.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="WindowSplit">
<div class="nav-panel">
<p>
Previous: <a href="#RenameChr" accesskey="p" rel="prev">Rename chromosomes in each block</a>, Up: <a href="#Extracting" accesskey="u" rel="up">Extracting data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Split-alignment-blocks-into-smaller-parts"><span>2.1.9 Split alignment blocks into smaller parts<a class="copiable-link" href="#Split-alignment-blocks-into-smaller-parts"> &para;</a></span></h4>

<p>The <code class="command">WindowSplit</code> filter subdivides blocks into smaller ones of a regular size, typically for window-based analyses.
Overlapping windows are partially supported (see options below).
</p>
<h3 class="heading" id="Synopsis_003a-9"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-9"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    WindowSplit(                            \
        preferred_size=1000,                \
        window_step=1000,                   \
        align=center,                       \
        keep_small_blocks=no),              \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-8"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-8"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">preferred_size={int&gt;0}</code></dt>
<dd><p>The preferred size for the output windows.
</p>
</dd>
<dt><code class="command">window_step={int&gt;0}</code></dt>
<dd><p>The nuùber of nucleotides by which the window should be slid. Only supported with &quot;align=ragged_left&quot;. Default value is equal to &quot;preferred_size&quot;, that is, non-overlapping windows.
</p>
</dd>
<dt><code class="command">align={ragged_left|ragged_right|center|adjust}</code></dt>
<dd><p>Specifies how to align windows on the block:
</p>
</dd>
<dt><code class="command">keep_small_blocks = {boolean}</code></dt>
<dd><p>Tell if blocks with an original size smaller that the window size should be kept (and untouched).
This option only works if the align option is set to adjust. The combination of &quot;align=adjust&quot; and &quot;keep_small_blocks=yes&quot; allows to ensure that blocks do not exceed a given size, without losing any data.
</p>
<dl class="table">
<dt><code class="command">ragged_left</code></dt>
<dd><pre class="verbatim">Block   |----------------------------|
Windows |----|----|----|----|----|XXXX
</pre>
</dd>
<dt><code class="command">ragged_right</code></dt>
<dd><pre class="verbatim">Block   |----------------------------|
Windows XXXX|----|----|----|----|----|
</pre>
</dd>
<dt><code class="command">center</code></dt>
<dd><pre class="verbatim">Block   |----------------------------|
Windows XX|----|----|----|----|----|XX
</pre>
</dd>
<dt><code class="command">adjust</code></dt>
<dd><pre class="verbatim">Block   |----------------------------|
Windows |-----|-----|-----|-----|----|
</pre></dd>
</dl>
<p>In the first three case, some parts of the original block (shown with X) will be discarded.
In the last case, the windows will have at least the specified size, but might be larger.
</p></dd>
</dl>


<hr>
</div>
</div>
<div class="section-level-extent" id="Cleaning">
<div class="nav-panel">
<p>
Next: <a href="#Analyzing" accesskey="n" rel="next">Statistical analysis</a>, Previous: <a href="#Extracting" accesskey="p" rel="prev">Extracting data</a>, Up: <a href="#Filters" accesskey="u" rel="up">Filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="Cleaning-alignment-blocks"><span>2.2 Cleaning alignment blocks<a class="copiable-link" href="#Cleaning-alignment-blocks"> &para;</a></span></h3>



<ul class="mini-toc">
<li><a href="#MinBlockLength" accesskey="1">Minumum block length</a></li>
<li><a href="#MinBlockSize" accesskey="2">Minumum block size</a></li>
<li><a href="#MinAlignmentScore" accesskey="3">Minumum alignment score</a></li>
<li><a href="#XFullGap" accesskey="4">Exclude full gap columns</a></li>
<li><a href="#RemoveEmptySequences" accesskey="5">Remove empty sequences</a></li>
<li><a href="#AlnFilter" accesskey="6">Alignment filtering</a></li>
<li><a href="#AlnFilter2" accesskey="7">Alignment filtering 2</a></li>
<li><a href="#EntropyFilter" accesskey="8">Entropy filtering</a></li>
<li><a href="#MaskFilter" accesskey="9">Masked sequences filtering</a></li>
<li><a href="#QualFilter">Quality filtering</a></li>
<li><a href="#FeatureFilter">Remove sequences from features from the alignment</a></li>
</ul>
<hr>
<div class="subsection-level-extent" id="MinBlockLength">
<div class="nav-panel">
<p>
Next: <a href="#MinBlockSize" accesskey="n" rel="next">Minumum block size</a>, Previous: <a href="#Cleaning" accesskey="p" rel="prev">Cleaning alignment blocks</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Minumum-block-length"><span>2.2.1 Minumum block length<a class="copiable-link" href="#Minumum-block-length"> &para;</a></span></h4>

<p>The <code class="command">MinBlockLength</code> filter discards blocks with less sites than a given threshold.
</p>
<h3 class="heading" id="Synopsis_003a-10"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-10"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    MinBlockLength(min_length=1000),        \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-9"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-9"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">min_length={int&gt;0}</code></dt>
<dd><p>The minimum length.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="MinBlockSize">
<div class="nav-panel">
<p>
Next: <a href="#MinAlignmentScore" accesskey="n" rel="next">Minumum alignment score</a>, Previous: <a href="#MinBlockLength" accesskey="p" rel="prev">Minumum block length</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Minumum-block-size"><span>2.2.2 Minumum block size<a class="copiable-link" href="#Minumum-block-size"> &para;</a></span></h4>

<p>The <code class="command">MinBlockSize</code> filter discards blocks with less sequences than a given threshold.
</p>
<h3 class="heading" id="Synopsis_003a-11"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-11"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    MinBlockSize(min_size=3),               \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-10"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-10"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">min_size={int&gt;0}</code></dt>
<dd><p>The minimum size.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="MinAlignmentScore">
<div class="nav-panel">
<p>
Next: <a href="#XFullGap" accesskey="n" rel="next">Exclude full gap columns</a>, Previous: <a href="#MinBlockSize" accesskey="p" rel="prev">Minumum block size</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Minumum-alignment-score"><span>2.2.3 Minumum alignment score<a class="copiable-link" href="#Minumum-alignment-score"> &para;</a></span></h4>

<p>The <code class="command">MinAlingmentScore</code> filter discards blocks with a score lower than a given threshold.
</p>
<h3 class="heading" id="Synopsis_003a-12"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-12"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    MinAlignmentScore(min_score=0.9),       \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-11"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-11"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">min_score={float&gt;0}</code></dt>
<dd><p>The minimum alignment score.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="XFullGap">
<div class="nav-panel">
<p>
Next: <a href="#RemoveEmptySequences" accesskey="n" rel="next">Remove empty sequences</a>, Previous: <a href="#MinAlignmentScore" accesskey="p" rel="prev">Minumum alignment score</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Exclude-full-gap-columns"><span>2.2.4 Exclude full gap columns<a class="copiable-link" href="#Exclude-full-gap-columns"> &para;</a></span></h4>

<p>Remove gap-only columns from blocks. This does not modify coordinates and might be necessary after a <code class="command">Subset</code> filter.
</p>
<h3 class="heading" id="Synopsis_003a-13"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-13"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    XFullGap(
        species=(species1,species2,etc)),   \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-12"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-12"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. Any column containing a gap in the specified species will be removed. If other species are present in the block, the corresponding character will be removed, whether it is a gap or not. The coordinates of these species will then be discarded.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="RemoveEmptySequences">
<div class="nav-panel">
<p>
Next: <a href="#AlnFilter" accesskey="n" rel="next">Alignment filtering</a>, Previous: <a href="#XFullGap" accesskey="p" rel="prev">Exclude full gap columns</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Remove-empty-sequences"><span>2.2.5 Remove empty sequences<a class="copiable-link" href="#Remove-empty-sequences"> &para;</a></span></h4>

<p>The <code class="command">RemoveEmptySequences</code> filter remove sequences containing only gap (or unresolved characters) in each block.
</p>
<h3 class="heading" id="Synopsis_003a-14"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-14"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    RemoveEmptySequences(unresolved_as_gaps=yes),        \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-13"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-13"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">unresolved_as_gaps={boolean}</code></dt>
<dd><p>Tell if unresolved characters count as &quot;empty&quot; as well.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="AlnFilter">
<div class="nav-panel">
<p>
Next: <a href="#AlnFilter2" accesskey="n" rel="next">Alignment filtering 2</a>, Previous: <a href="#RemoveEmptySequences" accesskey="p" rel="prev">Remove empty sequences</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Alignment-filtering"><span>2.2.6 Alignment filtering<a class="copiable-link" href="#Alignment-filtering"> &para;</a></span></h4>

<p>Split alignment blocks by removing regions with ambiguous alignments.
The local uncertainty in the alignment is determined through a sliding window based approach.
For each window, the number of gap characters and the total entropy are computed. Any window
for which both the entropy and number of gaps exceed the given thresholds will be removed from
the alignment, and the corresponding block split accordingly.
</p>
<h3 class="heading" id="Synopsis_003a-15"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-15"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    AlnFilter(                              \
        species=(species1,species2,etc),    \
        window_size=10,                     \
        window_step=1,                      \
        max_gap=9,                          \
        max_ent=0.2,                        \
        missing_as_gap=yes,                 \
        relative=no,                        \
        file=data.trash_aln.maf.gz,         \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-14"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-14"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. All calculations will be performed on the sub-alignment corresponding to these species only.
</p>
</dd>
<dt><code class="command">window_size={int&gt;0}</code></dt>
<dd><p>The width, in bp, of the sliding window.
</p>
</dd>
<dt><code class="command">window_step={int&gt;0}</code></dt>
<dd><p>The step by which the window is moved, in bp.
</p>
</dd>
<dt><code class="command">relative={boolean}</code></dt>
<dd><p>Tell if maximum amount of gap is relative (that is, as a proportion of the total amount of character in each window).
</p>
</dd>
<dt><code class="command">max.gap={int&gt;0|1&gt;double&gt;0}</code></dt>
<dd><p>The maximum number of gaps allowed in each window (if relative is set to no), or the maximum proportion of gaps (if relative is set to yes)
</p>
</dd>
<dt><code class="command">max_ent={float}</code></dt>
<dd><p>The maximum entropy allowed in each window.
</p>
</dd>
<dt><code class="command">missing_as_gap={yes/no}</code></dt>
<dd><p>Tell if missing sequences should be considered as gaps.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were removed alignment parts will be stored, in the MAF format.
This can be helpful for visual inspection and fine tuning of the filter parameters.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="AlnFilter2">
<div class="nav-panel">
<p>
Next: <a href="#EntropyFilter" accesskey="n" rel="next">Entropy filtering</a>, Previous: <a href="#AlnFilter" accesskey="p" rel="prev">Alignment filtering</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Alignment-filtering-2"><span>2.2.7 Alignment filtering 2<a class="copiable-link" href="#Alignment-filtering-2"> &para;</a></span></h4>

<p>This is another algorithm for cleaning alignment blocks (see AlnFilter), using sliding windows.
The number of gaps in each alignment column in the window is counted, and the column is masked if it contains more than a given threshold of gaps. consecutive patterns in the window are only counted ones. In the follwing 10nt window:
</p><pre class="verbatim">AATCGGGCGT
AA---GCGGA
AA---CGGGT
CA---CGGGA
</pre><p>positions 3, 4 and 5 will be masked if the maximum number of gaps allowed is 2 or less. The three columns will however count as only one indel event.
The window is then discarded if it contains more than a given number of indel events.
</p>
<h3 class="heading" id="Synopsis_003a-16"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-16"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    AlnFilter2(                             \
        species=(species1,species2,etc),    \
        window_size=10,                     \
        window_step=1,                      \
        max_gap=1,                          \
        max_pos=1,                          \
        relative=no,                        \
        missing_as_gap=yes,                 \
        file=data.trash_aln.maf.gz,         \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-15"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-15"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. All calculations will be performed on the sub-alignment corresponding to these species only.
</p>
</dd>
<dt><code class="command">window_size={int&gt;0}</code></dt>
<dd><p>The width, in bp, of the sliding window.
</p>
</dd>
<dt><code class="command">window_step={int&gt;0}</code></dt>
<dd><p>The step by which the window is moved, in bp.
</p>
</dd>
<dt><code class="command">relative={boolean}</code></dt>
<dd><p>Tell if maximum amount of gap is relative (that is, as a proportion of the total amount of character in each site).
</p>
</dd>
<dt><code class="command">max.gap={int&gt;0|1&gt;double&gt;0}</code></dt>
<dd><p>The maximum number of gaps allowed in each site (if relative is set to no), or the maximum proportion of gaps (if relative is set to yes)
</p>
</dd>
<dt><code class="command">max.pos={int&gt;0}</code></dt>
<dd><p>The maximum number of positions with gaps (&ldquo;indel events&rdquo;).
</p>
</dd>
<dt><code class="command">missing_as_gap={yes/no}</code></dt>
<dd><p>Tell if missing sequences should be considered as gaps.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were removed alignment parts will be stored, in the MAF format.
This can be helpful for visual inspection and fine tuning of the filter parameters.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="EntropyFilter">
<div class="nav-panel">
<p>
Next: <a href="#MaskFilter" accesskey="n" rel="next">Masked sequences filtering</a>, Previous: <a href="#AlnFilter2" accesskey="p" rel="prev">Alignment filtering 2</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Entropy-filtering"><span>2.2.8 Entropy filtering<a class="copiable-link" href="#Entropy-filtering"> &para;</a></span></h4>

<p>This filter removes highly variable regions and split the blocks accordingly.
It uses a sliding windows, and compute the entropy for each site in the window.
The window is then discarded if it containes more than &rsquo;p&rsquo; sites with an entropy higher that a user-specified threshold. The alignment block is then split into separate block accordingly.
</p>
<h3 class="heading" id="Synopsis_003a-17"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-17"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    EntropyFilter(                          \
        species=(species1,species2,etc),    \
        window_size=10,                     \
        window_step=1,                      \
        max_ent=0.2,                        \
        max_pos=3,                          \
        missing_as_gap=yes,                 \
        ignore_gaps=yes,                    \
        file=data.trash_ent.maf.gz,         \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-16"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-16"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. All calculations will be performed on the sub-alignment corresponding to these species only.
</p>
</dd>
<dt><code class="command">window_size={int&gt;0}</code></dt>
<dd><p>The width, in bp, of the sliding window.
</p>
</dd>
<dt><code class="command">window_step={int&gt;0}</code></dt>
<dd><p>The step by which the window is moved, in bp.
</p>
</dd>
<dt><code class="command">max_ent={float}</code></dt>
<dd><p>The maximum entropy allowed at each site.
</p>
</dd>
<dt><code class="command">max_pos={int&gt;0}</code></dt>
<dd><p>The maximum number of positions with an entropy higher than the given threshold.
</p>
</dd>
<dt><code class="command">missing_as_gap={yes/no}</code></dt>
<dd><p>Tell if unresolved characters should be counted as gaps.
</p>
</dd>
<dt><code class="command">ignore_gaps={yes/no}</code></dt>
<dd><p>Tell if gaps should not be counted in entropy calculation. If no, then gaps are counted as a &ldquo;fifth&rdquo; state.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were removed alignment parts will be stored, in the MAF format.
This can be helpful for visual inspection and fine tuning of the filter parameters.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="MaskFilter">
<div class="nav-panel">
<p>
Next: <a href="#QualFilter" accesskey="n" rel="next">Quality filtering</a>, Previous: <a href="#EntropyFilter" accesskey="p" rel="prev">Entropy filtering</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Masked-sequences-filtering"><span>2.2.9 Masked sequences filtering<a class="copiable-link" href="#Masked-sequences-filtering"> &para;</a></span></h4>

<p>The <code class="command">MaskFilter</code> split alignment blocks by removing regions with too many masked positions (typically showing repeat-content), using sliding windows.
A masked position is identified by a lower case letter in the original sequence.
Windows with more than a given amount of lower case characters will be discarded, and the corresponding block split.
</p>
<h3 class="heading" id="Synopsis_003a-18"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-18"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    MaskFilter(                             \
        species=(species1,species2,etc),    \
        window_size=10,                     \
        window_step=1,                      \
        max_masked=2,                       \
        file=data.trash_msk.maf.gz,         \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-17"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-17"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. All calculations will be performed on the sub-alignment corresponding to these species only.
</p>
</dd>
<dt><code class="command">window_size={int&gt;0}</code></dt>
<dd><p>The width, in bp, of the sliding window.
</p>
</dd>
<dt><code class="command">window_step={int&gt;0}</code></dt>
<dd><p>The step by which the window is moved, in bp.
</p>
</dd>
<dt><code class="command">max_masked={int&gt;0}</code></dt>
<dd><p>The maximum number of lower-case characters allowed in each window.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were removed alignment parts will be stored, in the MAF format.
This can be helpful for visual inspection and fine tuning of the filter parameters.
</p>
</dd>
<dt><code class="command">compresion={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="QualFilter">
<div class="nav-panel">
<p>
Next: <a href="#FeatureFilter" accesskey="n" rel="next">Remove sequences from features from the alignment</a>, Previous: <a href="#MaskFilter" accesskey="p" rel="prev">Masked sequences filtering</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Quality-filtering"><span>2.2.10 Quality filtering<a class="copiable-link" href="#Quality-filtering"> &para;</a></span></h4>

<p>The <code class="command">QualFilter</code> split alignment blocks by removing regions with too low quality score.
Windows with an average quality below a given threshold will be discarded, and the corresponding block split.
</p>
<h3 class="heading" id="Synopsis_003a-19"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-19"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    QualFilter(                             \
        species=(species1,species2,etc),    \
        window_size=10,                     \
        window_step=1,                      \
        min_qual=0.8,                       \
        file=data.trash_qual.maf.gz,        \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-18"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-18"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">species=(species1, species2, etc)</code></dt>
<dd><p>A comma-separated, within parentheses, list of species. All calculations will be performed on the sub-alignment corresponding to these species only. The corresponding sequences must have a quality score line in the original file.
</p>
</dd>
<dt><code class="command">window_size={int&gt;0}</code></dt>
<dd><p>The width, in bp, of the sliding window.
</p>
</dd>
<dt><code class="command">window_step={int&gt;0}</code></dt>
<dd><p>The step by which the window is moved, in bp.
</p>
</dd>
<dt><code class="command">min_qual={double}</code></dt>
<dd><p>The minimum average quality allowed for each window.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were removed alignment parts will be stored, in the MAF format.
This can be helpful for visual inspection and fine tuning of the filter parameters.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>

<hr>
</div>
<div class="subsection-level-extent" id="FeatureFilter">
<div class="nav-panel">
<p>
Previous: <a href="#QualFilter" accesskey="p" rel="prev">Quality filtering</a>, Up: <a href="#Cleaning" accesskey="u" rel="up">Cleaning alignment blocks</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Remove-sequences-from-features-from-the-alignment"><span>2.2.11 Remove sequences from features from the alignment<a class="copiable-link" href="#Remove-sequences-from-features-from-the-alignment"> &para;</a></span></h4>

<p>The <code class="command">FeatureFilter</code> remove parts of the alignment corresponding to certain features.
The features to extract are typically provided via a GFF file. See <code class="command">ExtractFeature</code>
in order to keep only the feature parts.
</p>
<h3 class="heading" id="Synopsis_003a-20"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-20"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    FeatureFilter(                          \
        ref_species=species1,               \
        feature.file=species1.gff3,         \
        feature.file.compression=none,      \
        feature.format=GFF,                 \
        feature.type=(mRNA,tRNA),           \
        file=data.trash_features.maf.gz,    \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-19"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-19"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">ref_species={string}</code></dt>
<dd><p>The name of the species for which the coordinates of the features are provided.
</p>
</dd>
<dt><code class="command">feature.file={path}</code></dt>
<dd><p>The file where the features are described.
</p>
</dd>
<dt><code class="command">feature.file.compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for the feature file.
</p>
</dd>
<dt><code class="command">feature.format={GFF|GTF|BedGraph}</code></dt>
<dd><p>Format for the feature file, currently GFF (v3.0), GTF or BedGraph.
</p>
</dd>
<dt><code class="command">feature.type={all|(list)}</code></dt>
<dd><p>Specifies which type of feature should be extracted from the feature file, if several are available.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were extracted features will be stored, in the MAF format.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>




<hr>
</div>
</div>
<div class="section-level-extent" id="Analyzing">
<div class="nav-panel">
<p>
Next: <a href="#Exporting" accesskey="n" rel="next">Exporting blocks and data</a>, Previous: <a href="#Cleaning" accesskey="p" rel="prev">Cleaning alignment blocks</a>, Up: <a href="#Filters" accesskey="u" rel="up">Filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="Statistical-analysis"><span>2.3 Statistical analysis<a class="copiable-link" href="#Statistical-analysis"> &para;</a></span></h3>


<ul class="mini-toc">
<li><a href="#Descriptive" accesskey="1">Descriptive statistics</a></li>
<li><a href="#Phylogenetics" accesskey="2">Phylogenetics</a></li>
</ul>
<hr>
<div class="subsection-level-extent" id="Descriptive">
<div class="nav-panel">
<p>
Next: <a href="#Phylogenetics" accesskey="n" rel="next">Phylogenetics</a>, Previous: <a href="#Analyzing" accesskey="p" rel="prev">Statistical analysis</a>, Up: <a href="#Analyzing" accesskey="u" rel="up">Statistical analysis</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Descriptive-statistics"><span>2.3.1 Descriptive statistics<a class="copiable-link" href="#Descriptive-statistics"> &para;</a></span></h4>

<p>The section describes how to compute simple statistics for each alignment block.
For the sake of computer efficiency, several statistics can be computed simultaneously on each alignment block.
The results are written in a text file, with one line per block and one column per statistics.
In addition, the coordinates of the block will be reported according to a specified reference species.
The choice of statistics is specified by the user. Some of them output several results, which will appear each in a column in the output file.
</p>
<p>Descriptive statistics are computed through the <code class="command">SequenceStatistics</code> filter, which takes the following arguments:
</p>
<h3 class="heading" id="Synopsis_003a-21"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-21"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    SequenceStatistics(                     \
        statistics=(\                       \
            BlockLength,                    \
            AlnScore,                       \
            BlockCounts),                   \
        ref_species=species1,               \
        file=data.statistics.csv,           \
        sep=comma,                          \
        compression=none),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-20"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-20"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">statistics={list of statistics functions}</code></dt>
<dd><p>See below for the list of possible functions and their detailed description.
</p>
</dd>
<dt><code class="command">ref_species={string}</code></dt>
<dd><p>The species to use to report block coordinates in the output file. For block where the reference species is missing, NA will be output.
</p>
</dd>
<dt><code class="command">file={path}</code></dt>
<dd><p>A file path for the output file.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">sep={comma,tab,{char}}</code></dt>
<dd><p>Column delimiter. Comma, tab, or any character (do not use the comma character directly, as it would lead to a format error when parsing the file).
</p>
</dd>
</dl>

<p>The statistics to compute take the form of functions (just like filters themselves), which can potentially take arguments.
Here is the list of currently available statistical functions:
</p>

<ul class="mini-toc">
<li><a href="#BlockSize" accesskey="1">Size of alignment blocks</a></li>
<li><a href="#BlockLength" accesskey="2">Length of alignment blocks</a></li>
<li><a href="#SequenceLength" accesskey="3">Length of sequence for a species</a></li>
<li><a href="#AlnScore" accesskey="4">Alignment score</a></li>
<li><a href="#BlockCounts" accesskey="5">Character frequencies</a></li>
<li><a href="#PairwiseDivergence" accesskey="6">Pairwise divergence</a></li>
<li><a href="#SiteStatistics" accesskey="7">Site statistics</a></li>
<li><a href="#FourSpeciesSitePatternCounts" accesskey="8">Counts of parsimony informative sites for a species quartet</a></li>
<li><a href="#SiteFrequencySpectrum" accesskey="9">Site frequency spectrum</a></li>
<li><a href="#PolymorphismStatistics">Polymorphism statistics</a></li>
<li><a href="#FstStatistics">Fst statistics</a></li>
<li><a href="#DiversityStatistics">Sequence diversity statistics</a></li>
<li><a href="#CountClusters">Number of sequence clusters.</a></li>
<li><a href="#ModelFit">Fit a substitution model and estimate parameters</a></li>
</ul>
<hr>
<div class="subsubsection-level-extent" id="BlockSize">
<div class="nav-panel">
<p>
Next: <a href="#BlockLength" accesskey="n" rel="next">Length of alignment blocks</a>, Previous: <a href="#Descriptive" accesskey="p" rel="prev">Descriptive statistics</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Size-of-alignment-blocks"><span>2.3.1.1 Size of alignment blocks<a class="copiable-link" href="#Size-of-alignment-blocks"> &para;</a></span></h4>
<p>The <code class="command">BlockSize</code> statistics reports the number of sequences in the block.
</p>
<p>(No argument)
</p>
<hr>
</div>
<div class="subsubsection-level-extent" id="BlockLength">
<div class="nav-panel">
<p>
Next: <a href="#SequenceLength" accesskey="n" rel="next">Length of sequence for a species</a>, Previous: <a href="#BlockSize" accesskey="p" rel="prev">Size of alignment blocks</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Length-of-alignment-blocks"><span>2.3.1.2 Length of alignment blocks<a class="copiable-link" href="#Length-of-alignment-blocks"> &para;</a></span></h4>
<p>The <code class="command">BlockLength</code> statistics reports the number of sites in the block.
</p>
<p>(No argument)
</p>
<hr>
</div>
<div class="subsubsection-level-extent" id="SequenceLength">
<div class="nav-panel">
<p>
Next: <a href="#AlnScore" accesskey="n" rel="next">Alignment score</a>, Previous: <a href="#BlockLength" accesskey="p" rel="prev">Length of alignment blocks</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Length-of-sequence-for-a-species"><span>2.3.1.3 Length of sequence for a species<a class="copiable-link" href="#Length-of-sequence-for-a-species"> &para;</a></span></h4>
<p>The <code class="command">SequenceLength</code> statistics computes the length (in bp) of the sequence of one species. It returns 0 if no sequence for this species is found in the current block.
It will send an error however if the block has more than one sequence for this species (be sure to use the <code class="command">Subset</code> filter before, if needed).
</p>
<h3 class="heading" id="Synopsis_003a-22"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-22"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            SequenceLength(                 \
                species=species1),          \
            [...]),                         \
        ref_species=species1,               \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-21"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-21"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species={string}</code></dt>
<dd><p>The species for which lengths should be reported.
</p></dd>
</dl>

<hr>
</div>
<div class="subsubsection-level-extent" id="AlnScore">
<div class="nav-panel">
<p>
Next: <a href="#BlockCounts" accesskey="n" rel="next">Character frequencies</a>, Previous: <a href="#SequenceLength" accesskey="p" rel="prev">Length of sequence for a species</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Alignment-score"><span>2.3.1.4 Alignment score<a class="copiable-link" href="#Alignment-score"> &para;</a></span></h4>
<p>The <code class="command">AlnScore</code> statistics reports the alignment score associated to the block, if any.
</p>
<p>(No argument)
</p>
<hr>
</div>
<div class="subsubsection-level-extent" id="BlockCounts">
<div class="nav-panel">
<p>
Next: <a href="#PairwiseDivergence" accesskey="n" rel="next">Pairwise divergence</a>, Previous: <a href="#AlnScore" accesskey="p" rel="prev">Alignment score</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Character-frequencies"><span>2.3.1.5 Character frequencies<a class="copiable-link" href="#Character-frequencies"> &para;</a></span></h4>
<p>The <code class="command">BlockCounts</code> statistics reports the count of each character found in the block.
</p>
<p>Note: this statistics does not work in case of duplicated sequences (multiple hits from the same chromosome / scaffold / contig).
</p>
<h3 class="heading" id="Arguments_003a-22"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-22"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species={list}</code></dt>
<dd><p>A list of species to be considered in the counts calculations. If no species is given, all sequences are used.
</p></dd>
<dt><code class="command">suffix={string}</code></dt>
<dd><p>An (optional) suffix to append to the statistics name. Useful in case the statistics should be computed for several species independently.
</p></dd>
</dl>

<hr>
</div>
<div class="subsubsection-level-extent" id="PairwiseDivergence">
<div class="nav-panel">
<p>
Next: <a href="#SiteStatistics" accesskey="n" rel="next">Site statistics</a>, Previous: <a href="#BlockCounts" accesskey="p" rel="prev">Character frequencies</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Pairwise-divergence"><span>2.3.1.6 Pairwise divergence<a class="copiable-link" href="#Pairwise-divergence"> &para;</a></span></h4>
<p>The <code class="command">PairwiseDivergence</code> computes the divergence (percentage of mistmatch) between two species.
If at least one of the species is missing, NaN is returned.
It will send an error however if the block has more than one sequence for this species (be sure to use the <code class="command">Subset</code> filter before, if needed).
</p>
<h3 class="heading" id="Synopsis_003a-23"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-23"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            PairwiseDivergence(             \
                species1=species1,          \
                species2=species1),         \
            [...]),                         \
        ref_species=species1,               \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-23"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-23"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species1={string}</code></dt>
<dd><p>The first species for which divergence should be reported.
</p></dd>
<dt><code class="command">species2={string}</code></dt>
<dd><p>The second species for which divergence should be reported.
</p></dd>
</dl>

<hr>
</div>
<div class="subsubsection-level-extent" id="SiteStatistics">
<div class="nav-panel">
<p>
Next: <a href="#FourSpeciesSitePatternCounts" accesskey="n" rel="next">Counts of parsimony informative sites for a species quartet</a>, Previous: <a href="#PairwiseDivergence" accesskey="p" rel="prev">Pairwise divergence</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Site-statistics"><span>2.3.1.7 Site statistics<a class="copiable-link" href="#Site-statistics"> &para;</a></span></h4>
<p><code class="command">SiteStatistics</code> computes site-wise statistics, including:
</p><ul class="itemize mark-bullet">
<li>Number of sites without gap,
</li><li>Number of complete sites (no gap, no unresolved character),
</li><li>Number of complete sites with only one state (constant sites),
</li><li>Number of complete sites with only two states (biallelic sites),
</li><li>Number of complete sites with only three states (triallelic sites),
</li><li>Number of complete sites with all four states (quadriallelic sites),
</li><li>Number of parsimony-informative sites.
</li></ul>
<p>The statistics are computed for a given subset of species (typically forming a ingroup).
</p>
<h3 class="heading" id="Synopsis_003a-24"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-24"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            SiteStatistics(                 \
                species=species1),          \
            [...]),                         \
        ref_species=species1,               \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-24"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-24"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species={list}</code></dt>
<dd><p>A list of species to be considered in the statistics calculations.
</p></dd>
</dl>

<hr>
</div>
<div class="subsubsection-level-extent" id="FourSpeciesSitePatternCounts">
<div class="nav-panel">
<p>
Next: <a href="#SiteFrequencySpectrum" accesskey="n" rel="next">Site frequency spectrum</a>, Previous: <a href="#SiteStatistics" accesskey="p" rel="prev">Site statistics</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Counts-of-parsimony-informative-sites-for-a-species-quartet"><span>2.3.1.8 Counts of parsimony informative sites for a species quartet<a class="copiable-link" href="#Counts-of-parsimony-informative-sites-for-a-species-quartet"> &para;</a></span></h4>
<p>The <code class="command">FourSpeciesSitePatternCounts</code> computes the frequency of each type of informative sites.
With four species, 3 types are expected: 1100, 0110 and 1010. On entry will be output for each of these types.
The remaining sites will be pulled in a column called &ldquo;Ignored&rdquo;.
</p> 
<h3 class="heading" id="Synopsis_003a-25"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-25"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            FourSpeciesSitePatternCounts(   \
                species1=sp1,               \
                species2=sp2,               \
                species3=sp3,               \
                species4=sp4                \
            [...]),                         \
        ref_species=sp1,                    \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-25"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-25"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species1=species</code></dt>
<dd><p>First species to consider (resp. second, third and fourth).
</p>
</dd>
</dl>

<hr>
</div>
<div class="subsubsection-level-extent" id="SiteFrequencySpectrum">
<div class="nav-panel">
<p>
Next: <a href="#PolymorphismStatistics" accesskey="n" rel="next">Polymorphism statistics</a>, Previous: <a href="#FourSpeciesSitePatternCounts" accesskey="p" rel="prev">Counts of parsimony informative sites for a species quartet</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Site-frequency-spectrum"><span>2.3.1.9 Site frequency spectrum<a class="copiable-link" href="#Site-frequency-spectrum"> &para;</a></span></h4>
<p>The <code class="command">SiteFrequencySpectrum</code> computes the site frequency spectrum for each block.
Only positions in the alignment with only two states are considered.
The proportions of bi-allelic sites are then computed by bins.
Lets consider the following example with 7 sequences:
</p><pre class="verbatim">ACGT
ACTT
AGTT
AGTT
TCTT
TCTT
TCTT
</pre><p>It contains one site with 4/7 (eq 3/7), one site with 2/7 (eq 5/7), one site with 1/7 (eq 6/7) and one site 0/7 (eq 7/7).
With seven sequences, there are actually 4 possibles frequencies: 1/7, 2/7, 3/7 and 4/7 (plus the 0/7 frequency, corresponding to a site with no-mutation).
It is possible to compute all these frequencies individually by settingthe <code class="command">bounds</code> parameter to
</p><pre class="verbatim">SiteFrequencySpectrum(bounds=(-0.5,0.5,1.5,2.5,3.5,4.5), ...)
</pre><p>which will output the number of site with 0, 1, 2, 3 or 4 minor states. Sites with more than 2 states are always counted separately, as well as sites containing unresolved characters.
If one want to count only constant sites for instance, one can simply type
</p><pre class="verbatim">SiteFrequencySpectrum(bounds=(-0.5,0.5), ...)
</pre><p>The remaining sites will be pulled in a column called &ldquo;Ignored&rdquo;.
</p> 
<h3 class="heading" id="Synopsis_003a-26"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-26"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            SiteFrequencySpectrum(          \
                bounds=(-0.5, 0.5, 1.5),    \
                ingroup=(pop1, pop2, pop3), \
                outgroup=species2,          \
            [...]),                         \
        ref_species=pop1,                   \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-26"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-26"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">bounds={list of double}</code></dt>
<dd><p>The bounds delimiting the bins of the spectrum to be computed.
</p>
</dd>
<dt><code class="command">ingroup={list}</code></dt>
<dd><p>A list of species forming the ingroup and on which the statistics should be calculated.
</p>
</dd>
<dt><code class="command">outgroup=species</code></dt>
<dd><p>A species name saying which sequence should be used to determine the ancestral state. If non-empty, the unfolded frequency spectrum will be computed.
If empty, the folded frequency spectrum will be returned (1/5 and 4/5 frequences will be pulled, so that the maximum frequency will be 1/2).
</p>
</dd>
</dl>

<hr>
</div>
<div class="subsubsection-level-extent" id="PolymorphismStatistics">
<div class="nav-panel">
<p>
Next: <a href="#FstStatistics" accesskey="n" rel="next">Fst statistics</a>, Previous: <a href="#SiteFrequencySpectrum" accesskey="p" rel="prev">Site frequency spectrum</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Polymorphism-statistics"><span>2.3.1.10 Polymorphism statistics<a class="copiable-link" href="#Polymorphism-statistics"> &para;</a></span></h4>
<p>Compute various statistics describing sequence polymorphism. Two aligned sets of &ldquo;species&rdquo; are compared, and the number of polymorphic / fixed sites are computed: 
</p><dl class="table">
<dt><code class="command">F</code></dt>
<dd><p>Number of sites fixed in the two sets
</p></dd>
<dt><code class="command">FF</code></dt>
<dd><p>Number of sites fixed in the two sets, yet with a distinct state
</p></dd>
<dt><code class="command">P</code></dt>
<dd><p>Number of sites that are polymorphic in the two sets
</p></dd>
<dt><code class="command">FP</code></dt>
<dd><p>Number of sites that are fixed in set 1 but polymorhic in set 2
</p></dd>
<dt><code class="command">PF</code></dt>
<dd><p>Number of sites that are fixed in set 2 but polymorphic in set 1
</p></dd>
</dl>

<p>Positions containing a gap or an unresolved character in one set are considered ambiguous.
Such positions are counted separately in the following quantities:
</p><dl class="table">
<dt><code class="command">X</code></dt>
<dd><p>Number of sites unresolved in the two sets
</p></dd>
<dt><code class="command">FX</code></dt>
<dd><p>Number of sites fixed in set 1 and unresolved in set 2
</p></dd>
<dt><code class="command">XF</code></dt>
<dd><p>Number of sites fixed in set 2 and unresolved in set 1
</p></dd>
<dt><code class="command">PX</code></dt>
<dd><p>Number of sites polymorphic in set 1 and unresolved in set 2
</p></dd>
<dt><code class="command">XP</code></dt>
<dd><p>Number of sites polymorphic in set 2 and unresolved in set 1
</p></dd>
</dl>

<h3 class="heading" id="Synopsis_003a-27"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-27"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            PolymorphismStatistics(         \
                species1=(A,B,C,D),         \
                species2=(E,F,G,H,I)        \
            [...]),                         \
        ref_species=A,                      \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-27"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-27"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species1={list}</code></dt>
<dd><p>A list of species for set 1.
</p></dd>
<dt><code class="command">species2={list}</code></dt>
<dd><p>A list of species for set 2.
</p></dd>
</dl>

<p>Note that the &ldquo;species&rdquo; terminology relates to multispecies alignments, as originally implemented in the MultiZ aligner. These statistics will however be most relevant when the aligned sequences are actually from individuals from the same population / species. The term &ldquo;species&rdquo; is here therefore to be taken in terchnical terms (a sequence id in the alignment), and not biological. 
</p>
<hr>
</div>
<div class="subsubsection-level-extent" id="FstStatistics">
<div class="nav-panel">
<p>
Next: <a href="#DiversityStatistics" accesskey="n" rel="next">Sequence diversity statistics</a>, Previous: <a href="#PolymorphismStatistics" accesskey="p" rel="prev">Polymorphism statistics</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Fst-statistics"><span>2.3.1.11 Fst statistics<a class="copiable-link" href="#Fst-statistics"> &para;</a></span></h4>
<p>Compute population divergence statistics (Fst), between two aligned sets of individuals. Hudson&rsquo;s 1992  estimator is used.
</p>
<h3 class="heading" id="Synopsis_003a-28"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-28"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            FstStatistics(                  \
                species1=(A,B,C,D),         \
                species2=(E,F,G,H,I),       \
                permutation_number = 1000,  \
                verbose = yes               \
            [...]),                         \
        ref_species=A,                      \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-28"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-28"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">species1={list}</code></dt>
<dd><p>A list of species for set 1.
</p></dd>
<dt><code class="command">species2={list}</code></dt>
<dd><p>A list of species for set 2.
</p></dd>
<dt><code class="command">min_permutation_number={int&gt;=0}</code></dt>
<dd><p>The minimum number of permutations to perform to compute p-values. If set to 0 (the default), no p-value computation is performed.
</p></dd>
<dt><code class="command">max_permutation_number={int&gt;=0}</code></dt>
<dd><p>If no permutation led to a Fst value at least as high as the observed one, continue until at least one such value is reach, or the specified maximum number of permutations is reached.
This allows the minimum possible p-value to be 1/(max_permutation_number + 1) instead of 1/(min_permutation_number + 1), without spending to much time on cases where the Fst is not significant.
</p></dd>
<dt><code class="command">verbose={boolean}</code></dt>
<dd><p>Tell if a progress bar should be shown in case permutations are to be performed.
</p></dd>
</dl>

<p>Note that the &ldquo;species&rdquo; terminology relates to multispecies alignments, as originally implemented in the MultiZ aligner. These statistics will however be most relevant when the aligned sequences are actually from individuals from the same population / species. The term &ldquo;species&rdquo; is here therefore to be taken in terchnical terms (a sequence id in the alignment), and not biological. 
</p>


<hr>
</div>
<div class="subsubsection-level-extent" id="DiversityStatistics">
<div class="nav-panel">
<p>
Next: <a href="#CountClusters" accesskey="n" rel="next">Number of sequence clusters.</a>, Previous: <a href="#FstStatistics" accesskey="p" rel="prev">Fst statistics</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Sequence-diversity-statistics"><span>2.3.1.12 Sequence diversity statistics<a class="copiable-link" href="#Sequence-diversity-statistics"> &para;</a></span></h4>
<p><code class="command">DiversityStatistics</code> computes sequence diversity statistics, including:
</p><ul class="itemize mark-bullet">
<li>Number of segregating sites,
</li><li>Watterson&rsquo;s theta.
</li><li>Tajima&rsquo;s pi (average pairwise dissimilarity).
</li><li>Tajima&rsquo;s D.
</li></ul>

<h3 class="heading" id="Synopsis_003a-29"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-29"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                   \
    [...],                                   
    SequenceStatistics(                       \
        statistics=(\                         \
            [...],                                                    
            DiversityStatistics(              \
                ingroup=(samp1,samp2,samp3)), \
            [...]),                           \
        ref_species=samp1,                    \
        file=data.statistics.csv),            \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-29"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-29"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">ingroup={list}</code></dt>
<dd><p>A list of sequences to be considered in the statistics calculations.
</p></dd>
</dl>


<hr>
</div>
<div class="subsubsection-level-extent" id="CountClusters">
<div class="nav-panel">
<p>
Next: <a href="#ModelFit" accesskey="n" rel="next">Fit a substitution model and estimate parameters</a>, Previous: <a href="#DiversityStatistics" accesskey="p" rel="prev">Sequence diversity statistics</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Number-of-sequence-clusters_002e"><span>2.3.1.13 Number of sequence clusters.<a class="copiable-link" href="#Number-of-sequence-clusters_002e"> &para;</a></span></h4>
<p><code class="command">CountClusters</code> counts the number of sequence clusters. This requires that a clustering tree was previously computed (see below).
</p>
<h3 class="heading" id="Synopsis_003a-30"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-30"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            CountClusters(                  \
                tree=bionj,                 \
                threshold=0.001),           \
            [...]),                         \
        ref_species=species1,               \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-30"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-30"> &para;</a></span></h3>
<dl class="table">
<dt><code class="command">tree={string}</code></dt>
<dd><p>The tag name used to store the previously computed tree.
</p>
</dd>
<dt><code class="command">threshold={double}</code></dt>
<dd><p>The threshold to use in order to compute clusters. A threshold of 0.001 for instance means that sequences within each cluster will be divergent by maximum 0.001, while sequences in distinct clusters will be divergent by at least 0.001.
</p> 
</dd>
</dl>


<hr>
</div>
<div class="subsubsection-level-extent" id="ModelFit">
<div class="nav-panel">
<p>
Previous: <a href="#CountClusters" accesskey="p" rel="prev">Number of sequence clusters.</a>, Up: <a href="#Descriptive" accesskey="u" rel="up">Descriptive statistics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Fit-a-substitution-model-and-estimate-parameters"><span>2.3.1.14 Fit a substitution model and estimate parameters<a class="copiable-link" href="#Fit-a-substitution-model-and-estimate-parameters"> &para;</a></span></h4>

<p><code class="command">ModelFit</code> fits a substitution model to the input block, given a phylogenetic tree.
All nucleotide homogeneous models can be used, with or without rate across sites variation.
</p>
<h3 class="heading" id="Synopsis_003a-31"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-31"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],                                   
    SequenceStatistics(                     \
        statistics=(\                       \
            [...],                                                    
            ModelFit(                       \
                model=HKY85(                \
                     kappa=1,               \
                     theta=0.5,             \
                     theta1=0.5,            \
                     theta2=0.5),           \
                rate_distribution=Gamma(    \
                     n=4,                   \
                     alpha=0.5),            \
                tree=BioNJ,                 \
                parameters_output=(         \
                     HKY85.theta,           \
                     HKY85.theta1,          \
                     HKY85.theta2,          \
                     HKY85.kappa),          \
                fixed_parameters=(),        \
                reestimate_brlen=yes,       \
                max_freq_gaps=0.3,          \
                gaps_as_unresolved=yes)),   \
            [...]),                         \
        ref_species=species1,               \
        file=data.statistics.csv),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-31"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-31"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">model={string}</code></dt>
<dd><p>Substitution model to use. See the Bio++ Program Suite manual for an extensive description of available models.
All nucleotide models can be used.
</p>
</dd>
<dt><code class="command">rate_distribution={string}</code></dt>
<dd><p>The distribution for rates across sites. See the Bio++ Program Suite manual for all available distributions.
</p>
</dd>
<dt><code class="command">root_freq={None|Full|GC}</code></dt>
<dd><p>Allow root frequencies to be different (non-stationary model). Root frequencies can be fully parametrized, or parametrized with GC content.
</p>
</dd>
<dt><code class="command">tree={string|none}</code></dt>
<dd><p>The property name under which trees are stored for each block. If set to &ldquo;none&rdquo;, then an input file should be given.
</p>
</dd>
<dt><code class="command">tree.file={path}[tree=none]</code></dt>
<dd><p>Path for tree file, in case no property is set.
</p>
</dd>
<dt><code class="command">tree.format={Newick|Nhx}[tree=none]</code></dt>
<dd><p>Format for tree file, in case no property is set.
</p>
</dd>
<dt><code class="command">parameters_output={list}</code></dt>
<dd><p>A list of parameter names to output as statistics.
</p>
</dd>
<dt><code class="command">fixed_parameters={list}</code></dt>
<dd><p>A list of parameters which should not be optimized, but fixed to their initila values.
</p>
</dd>
<dt><code class="command">reestimate_brlen={boolean}</code></dt>
<dd><p>Tell if the branches of the tree should be reestimated for each block together with other model parameters.
</p>
</dd>
<dt><code class="command">global_clock={boolean}</code></dt>
<dd><p>Tell if a global molecular clock should be enforced (only the heights of the nodes are optimized).
</p>
</dd>
<dt><code class="command">max_freq_gaps={float}</code></dt>
<dd><p>The maximum proportion of gaps for a site to be included in the analysis.
</p>
</dd>
<dt><code class="command">gaps_as_unresolved={yes/no}</code></dt>
<dd><p>Tell if remaining gaps should be converted to &rsquo;N&rsquo; before likelihood computation. This should be &rsquo;yes&rsquo; unless you specify a substitution model which explicitely allows for gaps.
</p>
</dd>
<dt><code class="command">global_clock={yes/no}</code></dt>
<dd><p>Assume a global clock for branch lengths.
</p>
</dd>
<dt><code class="command">reparametrize={yes/no}</code></dt>
<dd><p>Transform parameters to remove constraints (can improve optimization, but is usually slower).
</p>
</dd>
</dl>


<hr>
</div>
</div>
<div class="subsection-level-extent" id="Phylogenetics">
<div class="nav-panel">
<p>
Previous: <a href="#Descriptive" accesskey="p" rel="prev">Descriptive statistics</a>, Up: <a href="#Analyzing" accesskey="u" rel="up">Statistical analysis</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Phylogenetics-1"><span>2.3.2 Phylogenetics<a class="copiable-link" href="#Phylogenetics-1"> &para;</a></span></h4>



<ul class="mini-toc">
<li><a href="#DistanceEstimation" accesskey="1">Distance matrix estimation</a></li>
<li><a href="#DistanceBasedPhylogeny" accesskey="2">Hierarchical clustering based on distances</a></li>
<li><a href="#ExternalTreeBuilding" accesskey="3">Call an external phylogenetic software to reconstruct tree</a></li>
<li><a href="#NewOutgroup" accesskey="4">Root trees</a></li>
<li><a href="#DropSpecies" accesskey="5">Remove a species from a tree</a></li>
<li><a href="#TreeFilter" accesskey="6">Filter blocks based on their associated trees</a></li>
</ul>
<hr>
<div class="subsubsection-level-extent" id="DistanceEstimation">
<div class="nav-panel">
<p>
Next: <a href="#DistanceBasedPhylogeny" accesskey="n" rel="next">Hierarchical clustering based on distances</a>, Previous: <a href="#Phylogenetics" accesskey="p" rel="prev">Phylogenetics</a>, Up: <a href="#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Distance-matrix-estimation"><span>2.3.2.1 Distance matrix estimation<a class="copiable-link" href="#Distance-matrix-estimation"> &para;</a></span></h4>

<p>Estimates a pairwise distance matrix.
</p>
<h3 class="heading" id="Synopsis_003a-32"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-32"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    DistanceEstimation(                     \
        method=count,                       \
        gap_option=no_double_gap,           \
        unresolved_as_gap=no,               \
        extended_names=yes),                \
    [...]
</pre></div>
<p>or
</p><div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    DistanceEstimation(                     \
        method=ml,                          \
        model=K80(kappa=2),                 \
        rate=Gamma(n=4, alpha=0.5),         \
        parameter_estimation=initial,       \
        max_freq_gaps=0.33,                 \
        gaps_as_unresolved=yes,             \
        profiler=none,                      \
        message_handler=none,               \
        extended_names=yes),                \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-32"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-32"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">method={count|ml}</code></dt>
<dd><p>Method used to estimate distance, either observed count or maximum likelihood estimate.
</p></dd>
</dl>

<p>Further arguments for the observed counts:
</p><dl class="table">
<dt><code class="command">gap_option={string}</code></dt>
<dd><p>Specifies how to deal with gaps:
</p><dl class="table">
<dt><code class="command">all</code></dt>
<dd><p>All positions are used. Gaps are considered as a fifth character.
</p></dd>
<dt><code class="command">no_full_gap</code></dt>
<dd><p>Positions only made of gaps in the alignment block are ignored. Alternatively, a gap in the two sequences is ocnsidered as a match (gap are a &ldquo;fifth&rdquo; charcater).
</p></dd>
<dt><code class="command">no_double_gap</code></dt>
<dd><p>For each pairwise comparison, positions where a gap is found in both sequences are ignored.
</p></dd>
<dt><code class="command">no_gap</code></dt>
<dd><p>For each pairwise comparison, any gap-containing position is ignored. This is the recommended option for building phylogenies.
</p></dd>
</dl>

</dd>
<dt><code class="command">unresolved_as_gap={yes|no}</code></dt>
<dd><p>Tell is unresolved characters should be treated as gaps (usually in order to be ignored).
</p>
</dd>
<dt><code class="command">extended_names={boolean}</code></dt>
<dd><p>Tell if sequence coordinates should be included in the sequence names stored in the output matrix.
</p></dd>
</dl>

<p>Further arguments for the ML method:
</p><dl class="table">
<dt><code class="command">model={substitution model description}</code></dt>
<dd><p>See the Bio++ Program Suite manual for a description of substitution models available. Only nucleotide models can be used.
</p><div class="example">
<pre class="example-preformatted">model=JC
model=K80(kappa=2)
model=T92(kappa=2, theta=0.5)
model=GTR
</pre></div>

</dd>
<dt><code class="command">rate={rate distribution description}</code></dt>
<dd><p>See the Bio++ Program Suite manual for a description of available options.
</p><div class="example">
<pre class="example-preformatted">rate=Constant
rate=Gamma(n=4, alpha=0.5)
</pre></div>

</dd>
<dt><code class="command">profiler={none|std|{path}}</code></dt>
<dd><p>Where to print optimization steps (nowhere, standard output or to a given file).
</p>
</dd>
<dt><code class="command">message_handler={none|std|{path}}</code></dt>
<dd><p>Where to log optimization (nowhere, standard output or to a given file).
</p>
</dd>
<dt><code class="command">parameter_estimation={initial|pairwise}</code></dt>
<dd><p>How to estimate substitution process parameters (for instance kappa and alpha).
Available options are either to leave them to their initial values, or to estimate them for each pair of sequences.
</p>
</dd>
<dt><code class="command">max_freq_gaps={float}</code></dt>
<dd><p>The maximum proportion of gaps for a site to be included in the analysis.
</p>
</dd>
<dt><code class="command">gaps_as_unresolved={yes/no}</code></dt>
<dd><p>Tell if remaining gaps should be converted to &rsquo;N&rsquo; before likelihood computation. This should be &rsquo;yes&rsquo; unless you specify a substitution model which explicitely allows for gaps.
</p></dd>
</dl>


<h3 class="heading" id="Extra_002ddata_003a"><span>Extra-data:<a class="copiable-link" href="#Extra_002ddata_003a"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">CountDistance or MLDistance</code></dt>
<dd><p>The estimated pairwise distance matrix.
</p></dd>
</dl>


<hr>
</div>
<div class="subsubsection-level-extent" id="DistanceBasedPhylogeny">
<div class="nav-panel">
<p>
Next: <a href="#ExternalTreeBuilding" accesskey="n" rel="next">Call an external phylogenetic software to reconstruct tree</a>, Previous: <a href="#DistanceEstimation" accesskey="p" rel="prev">Distance matrix estimation</a>, Up: <a href="#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Hierarchical-clustering-based-on-distances"><span>2.3.2.2 Hierarchical clustering based on distances<a class="copiable-link" href="#Hierarchical-clustering-based-on-distances"> &para;</a></span></h4>

<p>Performs hierarchical clustering from a previously computed distance matrix.
</p>
<h3 class="heading" id="Synopsis_003a-33"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-33"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    DistanceBasedPhylogeny(                 \
        method=bionj,                       \
        dist_mat=MLDistance),               \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-33"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-33"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">method={string}</code></dt>
<dd><p>Method to use for reconstructing the tree:
</p><dl class="table">
<dt><code class="command">upgma</code></dt>
<dd><p>Unweighted pair group method using arithmetic means
</p></dd>
<dt><code class="command">wpgma</code></dt>
<dd><p>Weighted pair group method using aritmetic means
</p></dd>
<dt><code class="command">nj</code></dt>
<dd><p>Neighbor-Joining (Saitou and Nei, 1987)
</p></dd>
<dt><code class="command">bionj</code></dt>
<dd><p>BioNJ method (Gascuel, 1997)
</p></dd>
</dl>

</dd>
<dt><code class="command">dist_mat={string}</code></dt>
<dd><p>The tag name of the distance matrix to use.
</p></dd>
</dl>

<h3 class="heading" id="Extra_002ddata_003a-1"><span>Extra-data:<a class="copiable-link" href="#Extra_002ddata_003a-1"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">UPGMA|WPGMA|NJ|BioNJ</code></dt>
<dd><p>The resulting tree.
</p></dd>
</dl>


<hr>
</div>
<div class="subsubsection-level-extent" id="ExternalTreeBuilding">
<div class="nav-panel">
<p>
Next: <a href="#NewOutgroup" accesskey="n" rel="next">Root trees</a>, Previous: <a href="#DistanceBasedPhylogeny" accesskey="p" rel="prev">Hierarchical clustering based on distances</a>, Up: <a href="#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Call-an-external-phylogenetic-software-to-reconstruct-tree"><span>2.3.2.3 Call an external phylogenetic software to reconstruct tree<a class="copiable-link" href="#Call-an-external-phylogenetic-software-to-reconstruct-tree"> &para;</a></span></h4>

<p>Call an external phylogenetics software to build the tree and import it.
</p>
<h3 class="heading" id="Synopsis_003a-34"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-34"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    ExternalTreeBuilding(                   \
        input.file=tmp.phy,                 \
        input.format=Phylip(                \
          order=sequential, type=extended), \
        output.file=tmp.phy_phyml_tree,     \
        output.format=Newick,               \
        property_name=PhyML,                \
        call=runPhyml.sh,                   \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-34"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-34"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">input.file={none|{path}}</code></dt>
<dd><p>A file path where to write the block alignment, to be used as input by the external program.
</p>
</dd>
<dt><code class="command">input.format={string}</code></dt>
<dd><p>Alignment format for input file.
</p>
</dd>
<dt><code class="command">output.file={none|{path}}</code></dt>
<dd><p>A file path where to read the generated tree.
</p>
</dd>
<dt><code class="command">output.format={string}</code></dt>
<dd><p>The tree format.
</p>
</dd>
<dt><code class="command">call={string}</code></dt>
<dd><p>The system call to run the external program.
</p>
</dd>
<dt><code class="command">property_name={string}</code></dt>
<dd><p>The label to use when importing tree as meta-data.
</p>
</dd>
</dl>


<h3 class="heading" id="Extra_002ddata_003a-2"><span>Extra-data:<a class="copiable-link" href="#Extra_002ddata_003a-2"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">(user-specified tag name)</code></dt>
<dd><p>The reconstructed tree.
</p></dd>
</dl>



<hr>
</div>
<div class="subsubsection-level-extent" id="NewOutgroup">
<div class="nav-panel">
<p>
Next: <a href="#DropSpecies" accesskey="n" rel="next">Remove a species from a tree</a>, Previous: <a href="#ExternalTreeBuilding" accesskey="p" rel="prev">Call an external phylogenetic software to reconstruct tree</a>, Up: <a href="#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Root-trees"><span>2.3.2.4 Root trees<a class="copiable-link" href="#Root-trees"> &para;</a></span></h4>

<p>The <code class="command">NewOutgroup</code> filter takes an input tree and (re)root it according to a given species name.
</p>
<h3 class="heading" id="Synopsis_003a-35"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-35"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    NewOutgroup(                            \
        tree_input=BioNJ,                   \
        tree_output=NioNJ_rooted,           \
        outgroup=outgroupSpecies),          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-35"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-35"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">tree_input={string}</code></dt>
<dd><p>The tag name of the tree to reroot.
</p></dd>
<dt><code class="command">tree_output={string}</code></dt>
<dd><p>The tag name to use to store the rerooted tree. If the same as <code class="command">tree_input</code>, the rerooted tree will replace the old one.
</p></dd>
<dt><code class="command">outgroup={string}</code></dt>
<dd><p>The name of the species to use as outgroup. One and only one sequence for this species should be present in each block for which rerooting is to be performed.
</p></dd>
</dl>

<h3 class="heading" id="Extra_002ddata_003a-3"><span>Extra-data:<a class="copiable-link" href="#Extra_002ddata_003a-3"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">(user-defined tag)</code></dt>
<dd><p>A rooted tree.
</p></dd>
</dl>


<hr>
</div>
<div class="subsubsection-level-extent" id="DropSpecies">
<div class="nav-panel">
<p>
Next: <a href="#TreeFilter" accesskey="n" rel="next">Filter blocks based on their associated trees</a>, Previous: <a href="#NewOutgroup" accesskey="p" rel="prev">Root trees</a>, Up: <a href="#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Remove-a-species-from-a-tree"><span>2.3.2.5 Remove a species from a tree<a class="copiable-link" href="#Remove-a-species-from-a-tree"> &para;</a></span></h4>

<p>The <code class="command">DropSpecies</code> removes a leaf from a tree.
</p>
<h3 class="heading" id="Synopsis_003a-36"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-36"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    DropSpecies(                            \
        tree_input=BioNJ,                   \
        tree_output=BioNJ_subtree,          \
        species=species1),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-36"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-36"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">tree_input={string}</code></dt>
<dd><p>The tag name of the tree to edit.
</p></dd>
<dt><code class="command">tree_output={string}</code></dt>
<dd><p>The tag name to use to store the edited tree. If the same as <code class="command">tree_input</code>, the edited tree will replace the old one.
</p></dd>
<dt><code class="command">species={string}</code></dt>
<dd><p>The name of the species to remove. All leaves in the tree, if any, corresponding to a sequence for this species will be removed.
</p></dd>
</dl>

<h3 class="heading" id="Extra_002ddata_003a-4"><span>Extra-data:<a class="copiable-link" href="#Extra_002ddata_003a-4"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">(user-defined tag)</code></dt>
<dd><p>An edited tree.
</p></dd>
</dl>


<hr>
</div>
<div class="subsubsection-level-extent" id="TreeFilter">
<div class="nav-panel">
<p>
Previous: <a href="#DropSpecies" accesskey="p" rel="prev">Remove a species from a tree</a>, Up: <a href="#Phylogenetics" accesskey="u" rel="up">Phylogenetics</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsubsection" id="Filter-blocks-based-on-their-associated-trees"><span>2.3.2.6 Filter blocks based on their associated trees<a class="copiable-link" href="#Filter-blocks-based-on-their-associated-trees"> &para;</a></span></h4>

<p>The <code class="command">TreeFilter</code> (experimental) filters blocks based on their associated trees. Currently, filter is only performed based on the branch lengths.
</p>
<h3 class="heading" id="Synopsis_003a-37"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-37"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    TreeFilter(                             \
        tree=BioNJ,                         \
        max_brlen=0.01,                     \
        file=data.trash_tree.maf.gz,        \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-37"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-37"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">tree={string}</code></dt>
<dd><p>The tag name of the tree to analyse.
</p>
</dd>
<dt><code class="command">max_brlen={float&gt;0}</code></dt>
<dd><p>Branch length threshold: any block with a tree containing at least one branch lenght longer than the given threshold will be filtered out.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>An optional file were removed alignment parts will be stored, in the MAF format.
This can be helpful for visual inspection and fine tuning of the filter parameters.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file (if file != none).
</p>
</dd>
</dl>

<hr>
</div>
</div>
</div>
<div class="section-level-extent" id="Exporting">
<div class="nav-panel">
<p>
Next: <a href="#Miscelaneous" accesskey="n" rel="next">Miscelaneous filters</a>, Previous: <a href="#Analyzing" accesskey="p" rel="prev">Statistical analysis</a>, Up: <a href="#Filters" accesskey="u" rel="up">Filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="Exporting-blocks-and-data"><span>2.4 Exporting blocks and data<a class="copiable-link" href="#Exporting-blocks-and-data"> &para;</a></span></h3>


<ul class="mini-toc">
<li><a href="#Output" accesskey="1">Write alignment blocks to a MAF file</a></li>
<li><a href="#OutputAlignments" accesskey="2">Write alignment blocks to an external alignment file.</a></li>
<li><a href="#OutputAsTable" accesskey="3">Export given positions into a table format.</a></li>
<li><a href="#VcfOutput" accesskey="4">Call SNPs from alignment blocks and export to VCF file</a></li>
<li><a href="#MsmcOutput" accesskey="5">Export segregating sites to MSMC format</a></li>
<li><a href="#PlinkOutput" accesskey="6">Export segregating sites to PLINK format</a></li>
<li><a href="#OutputCoordinates" accesskey="7">Output coordinates of a set of species for each block.</a></li>
<li><a href="#OutputTrees" accesskey="8">Write trees to a file.</a></li>
<li><a href="#OutputDistanceMatrices" accesskey="9">Write distance matrices to a file.</a></li>
</ul>
<hr>
<div class="subsection-level-extent" id="Output">
<div class="nav-panel">
<p>
Next: <a href="#OutputAlignments" accesskey="n" rel="next">Write alignment blocks to an external alignment file.</a>, Previous: <a href="#Exporting" accesskey="p" rel="prev">Exporting blocks and data</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Write-alignment-blocks-to-a-MAF-file"><span>2.4.1 Write alignment blocks to a MAF file<a class="copiable-link" href="#Write-alignment-blocks-to-a-MAF-file"> &para;</a></span></h4>

<p>The <code class="command">Output</code> filter writes all blocks to a MAF file.
</p>
<h3 class="heading" id="Synopsis_003a-38"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-38"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    Output(                                 \
        file=data.filtered.maf.gz,          \
        compression=gzip,                   \
        mask=yes),                          \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-38"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-38"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path where to write alignment blocks, in the MAF format.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">mask={yes|no}</code></dt>
<dd><p>Tell if sequences should be masked (if a mask annotation is available), or if masking information should be ignored.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="OutputAlignments">
<div class="nav-panel">
<p>
Next: <a href="#OutputAsTable" accesskey="n" rel="next">Export given positions into a table format.</a>, Previous: <a href="#Output" accesskey="p" rel="prev">Write alignment blocks to a MAF file</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Write-alignment-blocks-to-an-external-alignment-file_002e"><span>2.4.2 Write alignment blocks to an external alignment file.<a class="copiable-link" href="#Write-alignment-blocks-to-an-external-alignment-file_002e"> &para;</a></span></h4>

<p>The <code class="command">OutputAlignments</code> filter writes all blocks to an external alignment file, potentially losing some information such as coordinates and scores.
Sequence names will be formated following the Ensembl convention when coordinates are available.
</p>
<h3 class="heading" id="Synopsis_003a-39"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-39"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    OutputAlignments(                       \
        file=data.filtered.aln,             \
        compression=none,                   \
        format=Clustal,                     \
        mask=no,                            \
        coordinates=yes),                   \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-39"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-39"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path where to write alignment blocks. If the file name contains the %i code, one alignment file per block will be generated, where %i will be replaced by the block index.
In addition to the %i special code, additional variables can be used (in combination with %i, in order to avoid putative identical file names): chromosome number (%c), begin (%b) and end (%e) of reference sequence.
</p>
</dd>
<dt><code class="command">format={string}</code></dt>
<dd><p>One of the alignment format supported by Bio++ (this includes Fasta, Mase and Clustal, see the Bio++ Program suite reference manual for a complete list of formats and options).
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">mask={yes|no}</code></dt>
<dd><p>Tell if sequences should be masked (if a mask annotation is available), or if masking information should be ignored.
</p>
</dd>
<dt><code class="command">coordinates={yes|no}</code></dt>
<dd><p>Tell if coordinates of sequences should be written in the header of each sequence (followinf Ensembl syntax).
If set to no, or if sequences have no coordinates, only the species name will be used as a sequence name.
</p>
</dd>
<dt><code class="command">ldhat_header={yes|no}</code></dt>
<dd><p>Tell if a header line should be added for each file, in order to be used with the LDhat convert program. This line contains the nimber of sequences, the sequence lengths, and the number 1 (for haploid).
</p>
</dd>
<dt><code class="command">reference={string}</code></dt>
<dd><p>The name of reference sequence, which will only be used when %c, %b or %e codes are used in file names. In case a block does not contain a sequence for the reference species, &quot;ChrNA&quot;, &quot;StartNA&quot; and &quot;StopNA&quot; tags will be used.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="OutputAsTable">
<div class="nav-panel">
<p>
Next: <a href="#VcfOutput" accesskey="n" rel="next">Call SNPs from alignment blocks and export to VCF file</a>, Previous: <a href="#OutputAlignments" accesskey="p" rel="prev">Write alignment blocks to an external alignment file.</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Export-given-positions-into-a-table-format_002e"><span>2.4.3 Export given positions into a table format.<a class="copiable-link" href="#Export-given-positions-into-a-table-format_002e"> &para;</a></span></h4>

<p>The <code class="command">OutputAsTable</code> filter writes the nucleotide content of given positions in a table format.
</p>
<h3 class="heading" id="Synopsis_003a-40"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-40"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    OutputTable(                            \
        file=data.txt,                      \
        species=Species1,Species2,Species3, \
        reference=RefSpecies,               \
        compression=none),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-40"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-40"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path where to write date in table format.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">species={list of {string}}</code></dt>
<dd><p>Select species for which data should be output (one column per species).
</p>
</dd>
<dt><code class="command">reference={string}</code></dt>
<dd><p>Which species should be used for outputting coordinates.
</p>
</dd>
</dl>



<hr>
</div>
<div class="subsection-level-extent" id="VcfOutput">
<div class="nav-panel">
<p>
Next: <a href="#MsmcOutput" accesskey="n" rel="next">Export segregating sites to MSMC format</a>, Previous: <a href="#OutputAsTable" accesskey="p" rel="prev">Export given positions into a table format.</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Call-SNPs-from-alignment-blocks-and-export-to-VCF-file"><span>2.4.4 Call SNPs from alignment blocks and export to VCF file<a class="copiable-link" href="#Call-SNPs-from-alignment-blocks-and-export-to-VCF-file"> &para;</a></span></h4>

<p>The <code class="command">VcfOutput</code> filter call SNPs from each block and output them to a VCF file.
Only substitutions are called for now.
</p>
<h3 class="heading" id="Synopsis_003a-41"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-41"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    VcfOutput(                              \
        file=snp.vcf.gz,                    \
        compression=gzip,                   \
        reference=Ref,                      \
        genotypes=(sample1,sample2),        \
        all=no,                             \
        diploids=no),                       \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-41"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-41"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path for the output VCF file.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">reference={species name}</code></dt>
<dd><p>A species name corresponding to the sequence to use as reference.
</p>
</dd>
<dt><code class="command">genotypes={list of species}</code></dt>
<dd><p>A list of species for which genotypes informations should be written. If not, void, a &rsquo;FORMAT&rsquo; column will be added to the output, as well as oneextra column per species to genotype.
If the species is not present in this particular, or if a N or Gap is present, a dot (&rsquo;.&rsquo;) will be written in the column.
Please note that species should be unique in each block, otherwise an error will occur.
Species can be further combined in groups as {(sample1,sample2),(sample3,sample4)} to create (phased) polyploids.
</p>
</dd>
<dt><code class="command">all={boolean}</code></dt>
<dd><p>If set to &rsquo;yes&rsquo;, also output non-variable positions in the VCF file.
</p>
</dd>
<dt><code class="command">diploids={boolean}</code></dt>
<dd><p>If set to &rsquo;yes&rsquo;, &rsquo;artificial&rsquo; homozygous diploids are written instead of haploid genotypes.
This option might be useful for downstream software that do not properly accomodate haploid data.
It only work when species are not already grouped in polyploids.
</p>
</dd>
<dt><code class="command">gap_as_deletion={boolean}</code></dt>
<dd><p>Gap characters are considered as a real genotype (deletion), rather than missing data.
This is currently done on a single-nucleotide basis, so that indels of length &gt; 1 are recorded as multiple entries.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="MsmcOutput">
<div class="nav-panel">
<p>
Next: <a href="#PlinkOutput" accesskey="n" rel="next">Export segregating sites to PLINK format</a>, Previous: <a href="#VcfOutput" accesskey="p" rel="prev">Call SNPs from alignment blocks and export to VCF file</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Export-segregating-sites-to-MSMC-format"><span>2.4.5 Export segregating sites to MSMC format<a class="copiable-link" href="#Export-segregating-sites-to-MSMC-format"> &para;</a></span></h4>

<p>The <code class="command">MsmcOutput</code> filter calls all segregating sites and export them to the MSMC format,
to be used with the MSMC program <a class="uref" href="https://github.com/stschiff/msmc">https://github.com/stschiff/msmc</a> (Multiple Sequentially Markovian Coalescent).
Sites with gap or unresolved characters are not called.
</p>
<h3 class="heading" id="Synopsis_003a-42"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-42"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    MsmcOutput(                             \
        file=snp.msmc,                      \
        compression=none,                   \
        reference=Ref,                      \
        genotypes=(sample1,sample2)),       \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-42"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-42"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path for the output MSMC file.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">reference={species name}</code></dt>
<dd><p>A species name corresponding to the sequence to use as reference.
</p>
</dd>
<dt><code class="command">genotypes={list of species}</code></dt>
<dd><p>A list of species for which segregating sites should be called. It might not contain the reference species, which is then only used for coordinates output.
Blocks with do not contain all species will not be called, as well as block not contianing the reference species. Note that the alignment has to be projected
againt the reference species, otherwise an error will be reported.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="PlinkOutput">
<div class="nav-panel">
<p>
Next: <a href="#OutputCoordinates" accesskey="n" rel="next">Output coordinates of a set of species for each block.</a>, Previous: <a href="#MsmcOutput" accesskey="p" rel="prev">Export segregating sites to MSMC format</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Export-segregating-sites-to-PLINK-format"><span>2.4.6 Export segregating sites to PLINK format<a class="copiable-link" href="#Export-segregating-sites-to-PLINK-format"> &para;</a></span></h4>

<p>The <code class="command">PlinkOutput</code> filter calls all biallelic segregating sites and export them to the PLINK format,
creating on Ped file and one Map file.
Sites with gap or unresolved characters are not called.
</p>
<h3 class="heading" id="Synopsis_003a-43"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-43"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    PlinkOutput(                            \
        ped_file=snp.ped,                   \
        map_file=snp.map,                   \
        compression=none,                   \
        reference=Ref,                      \
        genotypes=(sample1,sample2),        \
        recode_chr=no,                      \
        make_diploids=no,                   \
        map3=no,                            \
        phenotype=0,                        \
        tab_separator=yes)                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-43"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-43"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">ped_file={none|{path}}</code></dt>
<dd><p>A file path for the output Ped file.
</p>
</dd>
<dt><code class="command">map_file={none|{path}}</code></dt>
<dd><p>A file path for the output Map file.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">reference={species name}</code></dt>
<dd><p>A species name corresponding to the sequence to use as reference.
</p>
</dd>
<dt><code class="command">genotypes={list of species}</code></dt>
<dd><p>A list of species for which segregating sites should be called. It might not contain the reference species, which is then only used for coordinates output.
Blocks with do not contain all species will not be called, as well as block not contianing the reference species. Note that the alignment has to be projected
againt the reference species, otherwise an error will be reported.
</p>
</dd>
<dt><code class="command">recode_chr={boolean}</code></dt>
<dd><p>Tell if chromosome names should be converted to numbers. If yes, chromosomes will be numbered according to their order of occurrence in the MAF file.
More control over chromosome names can be achieved using the <code class="command">RenameChr</code> filter. 
</p>
</dd>
<dt><code class="command">make_diploids={boolean}</code></dt>
<dd><p>Tell if genomes should be combined into diploids. Genomes will be combined in the order given by the genotypes argument, genome1 with genome2, genome3 with genome4, etc. In case of an odd number of genomes, the last one will be ignored. If the option is set to &rsquo;no&rsquo;, each genome will be output as a homozygous diploid.
</p>
</dd>
<dt><code class="command">map3={boolean}</code></dt>
<dd><p>Tell if genetic distance column should be ommitted in the Map file. The resulting file should be sued together with the &ndash;map3 option in PLINK.
If set to no, then a 0 genetic distance is written.
</p>
</dd>
<dt><code class="command">phenotype={int}</code></dt>
<dd><p>The phenotype value, 6th column in the PED file. Usually 0 or -9 (missing value)
</p>
</dd>
<dt><code class="command">tab_separator={boolean}</code></dt>
<dd><p>Tell if tabulations should be used to separate columns and genotypes in the PED file. If set to &rsquo;no&rsquo;, then spaces will be used.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="OutputCoordinates">
<div class="nav-panel">
<p>
Next: <a href="#OutputTrees" accesskey="n" rel="next">Write trees to a file.</a>, Previous: <a href="#PlinkOutput" accesskey="p" rel="prev">Export segregating sites to PLINK format</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Output-coordinates-of-a-set-of-species-for-each-block_002e"><span>2.4.7 Output coordinates of a set of species for each block.<a class="copiable-link" href="#Output-coordinates-of-a-set-of-species-for-each-block_002e"> &para;</a></span></h4>

<p>The <code class="command">OutputCoordinates</code> takes as input a list of species and output the coordinates of each of the corresponding sequences in each block.
This filter only works if no more than one sequence per species are present, otherwise an error is casted. If some species are missing in some blocks, NA are produced in the output file.
</p>
<h3 class="heading" id="Synopsis_003a-44"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-44"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    OutputCoordinates(                      \
        file=coordinates.txt.gz,            \
        compression=gzip,                   \
        species=(species1,species2),        \
        output_src_size=yes)                \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-44"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-44"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path for the output text file.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">species={list of species}</code></dt>
<dd><p>A list of species for which coodinates should be written.
</p>
</dd>
<dt><code class="command">output_src_size={boolean}</code></dt>
<dd><p>Tell if the size of source sequences should be output.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="OutputTrees">
<div class="nav-panel">
<p>
Next: <a href="#OutputDistanceMatrices" accesskey="n" rel="next">Write distance matrices to a file.</a>, Previous: <a href="#OutputCoordinates" accesskey="p" rel="prev">Output coordinates of a set of species for each block.</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Write-trees-to-a-file_002e"><span>2.4.8 Write trees to a file.<a class="copiable-link" href="#Write-trees-to-a-file_002e"> &para;</a></span></h4>

<p>The <code class="command">OutputTrees</code> filter writes trees associated to blocks to a file, in the Newick format. All trees will be stored in the same file, one per line.
</p>
<h3 class="heading" id="Synopsis_003a-45"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-45"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    OutputTrees(                            \
        tree=BioNJ,                         \
        file=data.trees.dnd,                \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-45"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-45"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">tree={string}</code></dt>
<dd><p>The tag used to store previously computed trees.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path where to write trees, in the Newick format.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">strip_names={boolean}</code></dt>
<dd><p>Tell if leaf names should be stripped. If so, all text after the first dot will be deleted.
This is useful to get rid of the sequence coordinates.
The option will have no effect in case the names do not contain coordinates.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="OutputDistanceMatrices">
<div class="nav-panel">
<p>
Previous: <a href="#OutputTrees" accesskey="p" rel="prev">Write trees to a file.</a>, Up: <a href="#Exporting" accesskey="u" rel="up">Exporting blocks and data</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Write-distance-matrices-to-a-file_002e"><span>2.4.9 Write distance matrices to a file.<a class="copiable-link" href="#Write-distance-matrices-to-a-file_002e"> &para;</a></span></h4>

<p>The <code class="command">OutputDistanceMatrices</code> filter writes distance matrices associated to blocks to a file, in the Phylip format.
All matrices will be stored in the same file, one after the other.
</p>
<h3 class="heading" id="Synopsis_003a-46"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-46"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    OutputDistanceMatrices(                 \
        distance=MLdistance,                \
        file=data.distmat.ph,               \
        compression=gzip),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-46"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-46"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">distance={string}</code></dt>
<dd><p>The tag used to store previously computed distance matrices.
</p>
</dd>
<dt><code class="command">file={none|{path}}</code></dt>
<dd><p>A file path where to write distance matrices, in the Phylip format.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
<dt><code class="command">strip_names={boolean}</code></dt>
<dd><p>Tell if sequence names should be stripped. If so, all text after the first dot will be deleted.
This is useful to get rid of the sequence coordinates.
The option will have no effect in case the names do not contain coordinates.
</p>
</dd>
</dl>


<hr>
</div>
</div>
<div class="section-level-extent" id="Miscelaneous">
<div class="nav-panel">
<p>
Previous: <a href="#Exporting" accesskey="p" rel="prev">Exporting blocks and data</a>, Up: <a href="#Filters" accesskey="u" rel="up">Filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h3 class="section" id="Miscelaneous-filters"><span>2.5 Miscelaneous filters<a class="copiable-link" href="#Miscelaneous-filters"> &para;</a></span></h3>




<ul class="mini-toc">
<li><a href="#LiftOver" accesskey="1">Convert coordinates from one genome to another</a></li>
<li><a href="#SystemCall" accesskey="2">Call an external program on each block.</a></li>
</ul>
<hr>
<div class="subsection-level-extent" id="LiftOver">
<div class="nav-panel">
<p>
Next: <a href="#SystemCall" accesskey="n" rel="next">Call an external program on each block.</a>, Previous: <a href="#Miscelaneous" accesskey="p" rel="prev">Miscelaneous filters</a>, Up: <a href="#Miscelaneous" accesskey="u" rel="up">Miscelaneous filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Convert-coordinates-from-one-genome-to-another"><span>2.5.1 Convert coordinates from one genome to another<a class="copiable-link" href="#Convert-coordinates-from-one-genome-to-another"> &para;</a></span></h4>

<p>The <code class="command">LiftOver</code> mimmics the behavior of the famous UCSC liftover utility.
Features are read from any supported format, and a convertion table is generated for all features included in the alignment.
</p>
<h3 class="heading" id="Synopsis_003a-47"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-47"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...]
    LiftOver(                               \
        ref_species=species1,               \
        target_species=species2,            \
        target_closest_position=yes,        \
        feature.file=species1.gff3.gz,      \
        feature.file.compression=gzip,      \
        feature.format=GFF,                 \
        file=sp1_to_sp2.tln,                \
        compression=none),                  \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-47"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-47"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">ref_species={string}</code></dt>
<dd><p>The name of the species for which the coordinates of the features are provided.
</p>
</dd>
<dt><code class="command">target_species={string}</code></dt>
<dd><p>The name of the species to which the coordinates of the features should be converted.
</p>
</dd>
<dt><code class="command">target_closest_position={boolean}</code></dt>
<dd><p>In case the target sequence has a gap at the given position, outputs the coordinate of the previous non-gap position, otherwise NA.
</p>
</dd>
<dt><code class="command">feature.file={path}</code></dt>
<dd><p>The file where the features are described.
</p>
</dd>
<dt><code class="command">feature.file.compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for the feature file.
</p>
</dd>
<dt><code class="command">feature.format={GFF|GTF|BedGraph}</code></dt>
<dd><p>Format for the feature file, currently GFF (v3.0), GTF or BedGraph.
</p>
</dd>
<dt><code class="command">compression={none|gzip|zip|bzip2}</code></dt>
<dd><p>Compression format for output file.
</p>
</dd>
</dl>


<hr>
</div>
<div class="subsection-level-extent" id="SystemCall">
<div class="nav-panel">
<p>
Previous: <a href="#LiftOver" accesskey="p" rel="prev">Convert coordinates from one genome to another</a>, Up: <a href="#Miscelaneous" accesskey="u" rel="up">Miscelaneous filters</a> &nbsp; [<a href="#SEC_Contents" title="Table of contents" rel="contents">Contents</a>]</p>
</div>
<h4 class="subsection" id="Call-an-external-program-on-each-block_002e"><span>2.5.2 Call an external program on each block.<a class="copiable-link" href="#Call-an-external-program-on-each-block_002e"> &para;</a></span></h4>

<p>The <code class="command">SystemCall</code> filter runs an external program on each block.
For each block, the alignment is written to a temporary file, whose name and format can be specified.
This input file is then read by the external program. The called program is expected to generate results in the form of an alignment with the same sequence names, so that it can be read by maffilter.
In most case, the external program will be a wrapper script.
</p><h3 class="heading" id="Synopsis_003a-48"><span>Synopsis:<a class="copiable-link" href="#Synopsis_003a-48"> &para;</a></span></h3>

<table class="cartouche"><tr><td>
<div class="example">
<pre class="example-preformatted">maf.filter=                                 \
    [...],
    SystemCall(                             \
        name=myextprog,                     \
        input.file=tmp.fa,                  \
        input.format=Fasta,                 \
        call=myscript.sh,                   \
        output.file=output.fa,              \
        output.format=Fasta,                \
        hot=yes,                            \
        species_to_annotate=),              \
    [...]
</pre></div>
</td></tr></table>

<h3 class="heading" id="Arguments_003a-48"><span>Arguments:<a class="copiable-link" href="#Arguments_003a-48"> &para;</a></span></h3>

<dl class="table">
<dt><code class="command">name={string}</code></dt>
<dd><p>A label to use, might be useful in case several tools are called.
</p>
</dd>
<dt><code class="command">input.file={none|{path}}</code></dt>
<dd><p>A file path where to write the block alignment, to be used as input by the external program.
</p>
</dd>
<dt><code class="command">input.format={string}</code></dt>
<dd><p>Alignment format for input file.
</p>
</dd>
<dt><code class="command">output.file={none|{path}}</code></dt>
<dd><p>A file path where to read the processed block alignment, output by the external program.
</p>
</dd>
<dt><code class="command">output.format={string}</code></dt>
<dd><p>Alignment format for output file.
</p>
</dd>
<dt><code class="command">call={string}</code></dt>
<dd><p>The system call to run external program.
</p>
</dd>
<dt><code class="command">hot={bool}</code></dt>
<dd><p>Whether to perform a heads-or-tails test (align the reversed sequences and compare with the reversed alignment).
</p>
</dd>
<dt><code class="command">species_to_annotate={string}</code></dt>
<dd><p>if hot = yes and the argument is not empty, add position scores as annotations to the given species.
Use the first sequence for the species in case there are several. Annotations are not stored if no sequence for the species is available.
</p>

</dd>
</dl>






</div>
</div>
</div>
</div>



</body>
</html>
